// @flowgram.ai/form-materials/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  ArrayIcons: () => ArrayIcons,
  BatchVariableSelector: () => BatchVariableSelector,
  ConditionRow: () => ConditionRow,
  ConstantInput: () => ConstantInput,
  DynamicValueInput: () => DynamicValueInput,
  JsonSchemaEditor: () => JsonSchemaEditor,
  JsonSchemaUtils: () => JsonSchemaUtils,
  TypeSelector: () => TypeSelector,
  VariableSelector: () => VariableSelector,
  VariableTypeIcons: () => VariableTypeIcons,
  autoRenameRefEffect: () => autoRenameRefEffect,
  formatLegacyRefOnInit: () => formatLegacyRefOnInit,
  formatLegacyRefOnSubmit: () => formatLegacyRefOnSubmit,
  formatLegacyRefToNewRef: () => formatLegacyRefToNewRef,
  formatNewRefToLegacyRef: () => formatNewRefToLegacyRef,
  getSchemaIcon: () => getSchemaIcon,
  getTypeSelectValue: () => getTypeSelectValue,
  isLegacyFlowRefValueSchema: () => isLegacyFlowRefValueSchema,
  isNewFlowRefValueSchema: () => isNewFlowRefValueSchema,
  parseTypeSelectValue: () => parseTypeSelectValue,
  provideBatchInputEffect: () => provideBatchInputEffect,
  provideBatchOutputsEffect: () => provideBatchOutputsEffect
});
module.exports = __toCommonJS(src_exports);

// src/components/variable-selector/index.tsx
var import_react3 = __toESM(require("react"), 1);
var import_semi_icons2 = require("@douyinfe/semi-icons");

// src/components/variable-selector/use-variable-tree.tsx
var import_react2 = __toESM(require("react"), 1);
var import_editor2 = require("@flowgram.ai/editor");
var import_semi_ui = require("@douyinfe/semi-ui");

// src/components/type-selector/constants.tsx
var import_react = __toESM(require("react"), 1);
var import_semi_icons = __toESM(require("@douyinfe/semi-icons"), 1);
var VariableTypeIcons = {
  custom: /* @__PURE__ */ import_react.default.createElement(
    "svg",
    {
      viewBox: "0 0 24 24",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg",
      width: "1em",
      height: "1em",
      focusable: "false",
      "aria-hidden": "true"
    },
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        fillRule: "evenodd",
        clipRule: "evenodd",
        d: "M5.1 18L4.5032 20.1702C4.24999 21.0909 4.94281 22 5.89773 22C6.54881 22 7.11964 21.565 7.29227 20.9372L8.1 18H12.1L11.5032 20.1702C11.25 21.0909 11.9428 22 12.8977 22C13.5488 22 14.1196 21.565 14.2923 20.9372L15.1 18H19.5C20.3284 18 21 17.3284 21 16.5C21 15.6716 20.3284 15 19.5 15H15.925L17.575 9H20.5C21.3284 9 22 8.32843 22 7.5C22 6.67157 21.3284 6 20.5 6H18.4L18.9968 3.8298C19.25 2.90906 18.5572 2 17.6023 2C16.9512 2 16.3804 2.43504 16.2077 3.06281L15.4 6H11.4L11.9968 3.8298C12.25 2.90906 11.5572 2 10.6023 2C9.95119 2 9.38036 2.43504 9.20773 3.06281L8.4 6H4.5C3.67157 6 3 6.67157 3 7.5C3 8.32843 3.67157 9 4.5 9H7.575L5.925 15H3.5C2.67157 15 2 15.6716 2 16.5C2 17.3284 2.67157 18 3.5 18H5.1ZM8.925 15L10.575 9H14.575L12.925 15H8.925Z",
        fill: "currentColor"
      }
    )
  ),
  object: /* @__PURE__ */ import_react.default.createElement(
    "svg",
    {
      width: "1em",
      height: "1em",
      viewBox: "0 0 16 16",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg"
    },
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M5.33893 1.5835C5.66613 1.5835 5.93137 1.88142 5.93137 2.20862C5.93137 2.53582 5.66613 2.76838 5.33893 2.76838H4.9099C4.34717 2.76838 4.08062 3.07557 4.08062 3.71921V6.58633C4.08062 7.30996 3.80723 7.84734 3.26798 8.19105C3.11426 8.28902 3.10884 8.55273 3.26068 8.65359C3.80476 9.01503 4.08062 9.53994 4.08062 10.2434V13.1251C4.08062 13.7395 4.34717 14.0613 4.9099 14.0613H5.33893C5.66613 14.0613 5.93137 14.3435 5.93137 14.6707C5.93137 14.9979 5.66613 15.2462 5.33893 15.2462H4.64335C3.99177 15.2462 3.48828 15.0268 3.13287 14.6172C2.80708 14.2369 2.64419 13.7103 2.64419 13.0666V10.3165C2.64419 9.8923 2.55534 9.58511 2.37764 9.39494C2.26816 9.27135 1.80618 9.17938 1.38154 9.11602C1.02726 9.06315 0.759057 8.76744 0.765747 8.4093C0.772379 8.0543 1.03439 7.7566 1.38545 7.70346C1.80778 7.63952 2.26906 7.54968 2.37764 7.43477C2.55534 7.22997 2.64419 6.92278 2.64419 6.51319V3.77772C2.64419 3.11945 2.80708 2.59284 3.13287 2.21251C3.48828 1.78829 3.99177 1.5835 4.64335 1.5835H5.33893Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M10.962 15.2463C10.6348 15.2463 10.3696 14.9483 10.3696 14.6211C10.3696 14.2939 10.6348 14.0614 10.962 14.0614H11.391C11.9538 14.0614 12.2203 13.7542 12.2203 13.1105V10.2434C12.2203 9.51979 12.4937 8.98241 13.033 8.6387C13.1867 8.54073 13.1921 8.27703 13.0403 8.17616C12.4962 7.81472 12.2203 7.28982 12.2203 6.58638V3.70463C12.2203 3.09024 11.9538 2.76842 11.391 2.76842L10.962 2.76842C10.6348 2.76842 10.3696 2.48627 10.3696 2.15907C10.3696 1.83188 10.6348 1.58354 10.962 1.58354L11.6576 1.58354C12.3092 1.58354 12.8127 1.80296 13.1681 2.21255C13.4939 2.59289 13.6568 3.1195 13.6568 3.76314V6.51324C13.6568 6.93745 13.7456 7.24464 13.9233 7.43481C14.03 7.5553 14.4328 7.64858 14.8186 7.71393C15.1718 7.77376 15.4401 8.06977 15.4334 8.42791C15.4268 8.78291 15.1646 9.08018 14.814 9.13633C14.4306 9.19774 14.0291 9.28303 13.9233 9.39499C13.7456 9.59978 13.6568 9.90697 13.6568 10.3166V13.052C13.6568 13.7103 13.4939 14.2369 13.1681 14.6172C12.8127 15.0415 12.3092 15.2463 11.6576 15.2463H10.962Z",
        fill: "currentColor"
      }
    )
  ),
  boolean: /* @__PURE__ */ import_react.default.createElement(
    "svg",
    {
      width: "1em",
      height: "1em",
      viewBox: "0 0 16 16",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg"
    },
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        fillRule: "evenodd",
        clipRule: "evenodd",
        d: "M10.668 4.66683H5.33463C3.49369 4.66683 2.0013 6.15921 2.0013 8.00016C2.0013 9.84111 3.49369 11.3335 5.33463 11.3335H10.668C12.5089 11.3335 14.0013 9.84111 14.0013 8.00016C14.0013 6.15921 12.5089 4.66683 10.668 4.66683ZM5.33463 3.3335C2.75731 3.3335 0.667969 5.42283 0.667969 8.00016C0.667969 10.5775 2.75731 12.6668 5.33463 12.6668H10.668C13.2453 12.6668 15.3346 10.5775 15.3346 8.00016C15.3346 5.42283 13.2453 3.3335 10.668 3.3335H5.33463Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        fillRule: "evenodd",
        clipRule: "evenodd",
        d: "M8.66797 8.00016C8.66797 6.89559 9.5634 6.00016 10.668 6.00016C11.7725 6.00016 12.668 6.89559 12.668 8.00016C12.668 9.10473 11.7725 10.0002 10.668 10.0002C9.5634 10.0002 8.66797 9.10473 8.66797 8.00016ZM10.668 7.3335C10.2998 7.3335 10.0013 7.63197 10.0013 8.00016C10.0013 8.36835 10.2998 8.66683 10.668 8.66683C11.0362 8.66683 11.3346 8.36835 11.3346 8.00016C11.3346 7.63197 11.0362 7.3335 10.668 7.3335Z",
        fill: "currentColor"
      }
    )
  ),
  string: /* @__PURE__ */ import_react.default.createElement(
    "svg",
    {
      width: "1em",
      height: "1em",
      viewBox: "0 0 16 16",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg"
    },
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M9.3342 3.33321C8.96601 3.33321 8.66753 3.63169 8.66753 3.99988C8.66753 4.36807 8.96601 4.66655 9.3342 4.66655H14.6675C15.0357 4.66655 15.3342 4.36807 15.3342 3.99988C15.3342 3.63169 15.0357 3.33321 14.6675 3.33321H9.3342Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M10.0009 7.99988C10.0009 7.63169 10.2993 7.33321 10.6675 7.33321H14.6675C15.0357 7.33321 15.3342 7.63169 15.3342 7.99988C15.3342 8.36807 15.0357 8.66655 14.6675 8.66655H10.6675C10.2993 8.66655 10.0009 8.36807 10.0009 7.99988Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M12.0009 11.3332C11.6327 11.3332 11.3342 11.6317 11.3342 11.9999C11.3342 12.3681 11.6327 12.6665 12.0009 12.6665H14.6675C15.0357 12.6665 15.3342 12.3681 15.3342 11.9999C15.3342 11.6317 15.0357 11.3332 14.6675 11.3332H12.0009Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M9.86659 14.1482L8.23444 10.1844H3.18136C3.13868 10.1844 3.09685 10.1808 3.05616 10.1738L1.66589 14.1129C1.53049 14.4965 1.10971 14.6978 0.726058 14.5624C0.342408 14.427 0.141166 14.0062 0.276572 13.6225L4.37566 2.00848C4.71323 1.05202 6.05321 1.01763 6.4394 1.95552L11.2289 13.5872C11.3838 13.9634 11.2044 14.394 10.8282 14.5489C10.452 14.7038 10.0215 14.5244 9.86659 14.1482ZM5.44412 3.40791L3.57241 8.71109H7.62778L5.44412 3.40791Z",
        fill: "currentColor"
      }
    )
  ),
  integer: /* @__PURE__ */ import_react.default.createElement(
    "svg",
    {
      width: "1em",
      height: "1em",
      viewBox: "0 0 16 16",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg"
    },
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M15.132 11.4601C15.644 11.0121 15.9 10.3921 15.9 9.60007C15.9 8.60807 15.5 7.93607 14.7 7.58407C15.412 7.23207 15.768 6.62407 15.768 5.76007C15.768 5.05607 15.536 4.48007 15.072 4.03207C14.608 3.59207 14.012 3.37207 13.284 3.37207C12.588 3.37207 12.008 3.58007 11.544 3.99607C11.064 4.42007 10.808 4.98807 10.776 5.70007H12C12.064 4.88407 12.492 4.47607 13.284 4.47607C14.124 4.47607 14.544 4.91607 14.544 5.79607C14.544 6.66007 14.112 7.09207 13.248 7.09207H13.044V8.16007H13.248C14.2 8.16007 14.676 8.62807 14.676 9.56407C14.676 10.5081 14.212 10.9801 13.284 10.9801C12.9 10.9801 12.584 10.8761 12.336 10.6681C12.064 10.4441 11.916 10.1161 11.892 9.68407H10.668C10.692 10.4761 10.964 11.0841 11.484 11.5081C11.948 11.8921 12.548 12.0841 13.284 12.0841C14.036 12.0841 14.652 11.8761 15.132 11.4601Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M4.46875 12.0003V10.9083L7.75675 6.91228C8.06075 6.54428 8.21275 6.16428 8.21275 5.77228C8.21275 4.90828 7.79675 4.47628 6.96475 4.47628C6.60475 4.47628 6.31275 4.57628 6.08875 4.77628C5.83275 5.00828 5.70475 5.34828 5.70475 5.79628H4.48075C4.48075 5.07628 4.71275 4.49228 5.17675 4.04428C5.64075 3.60428 6.23675 3.38428 6.96475 3.38428C7.70075 3.38428 8.29675 3.60028 8.75275 4.03228C9.20875 4.47228 9.43675 5.05628 9.43675 5.78428C9.43675 6.13628 9.36875 6.45628 9.23275 6.74428C9.12075 6.97628 8.92075 7.27228 8.63275 7.63228L5.95675 10.9083H9.43675V12.0003H4.46875Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M1.668 12.0001V4.78805L0 6.25205V4.89605L1.668 3.45605H2.892V12.0001H1.668Z",
        fill: "currentColor"
      }
    )
  ),
  number: /* @__PURE__ */ import_react.default.createElement(
    "svg",
    {
      width: "1em",
      height: "1em",
      viewBox: "0 0 16 16",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg"
    },
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M3.44151 5.3068C3.44151 3.83404 4.71542 2.64014 6.18818 2.64014C7.66094 2.64014 8.93484 3.83404 8.93484 5.3068V10.6135C8.93484 12.0862 7.66094 13.2801 6.18818 13.2801C4.71542 13.2801 3.44151 12.0862 3.44151 10.6135V5.3068ZM7.60151 5.3068C7.60151 4.57042 6.92456 3.97347 6.18818 3.97347C5.4518 3.97347 4.77484 4.57042 4.77484 5.3068V10.6135C4.77484 11.3498 5.4518 11.9468 6.18818 11.9468C6.92456 11.9468 7.60151 11.3498 7.60151 10.6135V5.3068Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M12.9882 2.64014C11.5154 2.64014 10.2415 3.83404 10.2415 5.3068V10.6135C10.2415 12.0862 11.5154 13.2801 12.9882 13.2801C14.4609 13.2801 15.7348 12.0862 15.7348 10.6135V5.3068C15.7348 3.83404 14.4609 2.64014 12.9882 2.64014ZM14.4015 10.6135C14.4015 11.3498 13.7246 11.9468 12.9882 11.9468C12.2518 11.9468 11.5748 11.3498 11.5748 10.6135V5.3068C11.5748 4.57042 12.2518 3.97347 12.9882 3.97347C13.7246 3.97347 14.4015 4.57042 14.4015 5.3068V10.6135Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M1.21484 13.2001C1.76713 13.2001 2.21484 12.7524 2.21484 12.2001C2.21484 11.6479 1.76713 11.2001 1.21484 11.2001C0.662559 11.2001 0.214844 11.6479 0.214844 12.2001C0.214844 12.7524 0.662559 13.2001 1.21484 13.2001Z",
        fill: "currentColor"
      }
    )
  ),
  array: /* @__PURE__ */ import_react.default.createElement(
    "svg",
    {
      width: "1em",
      height: "1em",
      viewBox: "0 0 16 16",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg"
    },
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M5.23759 1.00342H2.00391V14.997H5.23759V13.6251H3.35127V2.37534H5.23759V1.00342Z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M10.7624 1.00342H13.9961V14.997H10.7624V13.6251H12.6487V2.37534H10.7624V1.00342Z",
        fill: "currentColor"
      }
    )
  ),
  stream: /* @__PURE__ */ import_react.default.createElement(
    "svg",
    {
      viewBox: "0 0 1024 1024",
      version: "1.1",
      xmlns: "http://www.w3.org/2000/svg",
      width: "1em",
      height: "1em"
    },
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M879.674 544.51l-158.254-0.221c-8.534 2.287-17.305-2.776-19.588-11.307l-23.862-75.877-74.742 350.891c0 0-1.523 18.507-11.518 18.507s-26.9 0.281-26.9 0.281c-8.259 2.213-16.748-2.687-18.961-10.949l-92.741-457.648-70.305 330.634c-2.261 8.291-11.94 15.206-20.385 12.986l-24.876 0.339c-8.723 2.293-17.685-2.789-20.023-11.349L270.629 544.51 143.993 544.51c-8.831 0-15.993-7.159-15.993-15.993l0-31.986c0-8.831 7.162-15.993 15.993-15.993l157.429-0.516c9.565-0.304 17.685 0.788 20.023 9.351l24.386 76.092 68.642-358.907c0 0 3.4-10.894 14.397-10.894 10.994 0 34.107-0.448 34.107-0.448 8.262-2.213 16.751 2.687 18.965 10.949l91.912 454.126 67.948-326.182c2.213-8.262 8.707-15.161 16.965-12.948l27.316-0.333c8.531-2.287 17.301 2.776 19.588 11.31l46.665 148.4 127.337 0c8.835 0 15.993 7.162 15.993 15.993l0 31.986C895.667 537.352 888.508 544.51 879.674 544.51z",
        fill: "currentColor"
      }
    )
  ),
  map: /* @__PURE__ */ import_react.default.createElement(
    "svg",
    {
      viewBox: "0 0 1024 1024",
      version: "1.1",
      xmlns: "http://www.w3.org/2000/svg",
      width: "1em",
      height: "1em"
    },
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M877.860571 938.642286h-645.851428c-27.574857 0-54.052571-11.337143-73.508572-31.744a110.957714 110.957714 0 0 1-30.500571-76.8V193.828571c0-28.745143 10.971429-56.32 30.500571-76.726857a101.888 101.888 0 0 1 73.508572-31.817143h574.171428c27.501714 0 53.979429 11.337143 73.508572 31.744 19.529143 20.333714 30.500571 48.054857 30.500571 76.8v522.020572a34.157714 34.157714 0 0 1-6.948571 22.820571c-37.156571 19.382857-57.636571 39.350857-57.636572 72.630857 0 39.716571 19.894857 50.029714 57.636572 72.777143a34.816 34.816 0 0 1-8.045714 49.298286 32.256 32.256 0 0 1-17.334858 5.193143z m-32.256-254.537143V193.828571a40.228571 40.228571 0 0 0-39.497142-41.179428H232.009143a40.301714 40.301714 0 0 0-39.497143 41.252571V699.245714c17.773714-9.874286 37.449143-14.994286 57.417143-14.921143h595.675428v-0.073142z m-595.675428 187.245714h566.198857c-22.893714-11.190857-27.940571-39.497143-28.013714-59.977143 0-20.260571 3.218286-43.885714 28.013714-59.904h-566.125714c-31.670857 0-57.417143 26.843429-57.417143 59.977143 0 33.060571 25.746286 59.904 57.344 59.904z",
        fill: "currentColor"
      }
    ),
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        d: "M320 128m32.036571 0l-0.073142 0q32.036571 0 32.036571 32.036571l0 511.926858q0 32.036571-32.036571 32.036571l0.073142 0q-32.036571 0-32.036571-32.036571l0-511.926858q0-32.036571 32.036571-32.036571Z",
        fill: "currentColor"
      }
    )
  )
};
var ArrayIcons = {
  object: /* @__PURE__ */ import_react.default.createElement(
    "svg",
    {
      width: "1em",
      height: "1em",
      viewBox: "0 0 16 16",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg"
    },
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        fillRule: "evenodd",
        clipRule: "evenodd",
        d: "M0 1.58105H3.6139V2.87326H1.36702V13.1264H3.6139V14.4186H0V1.58105ZM3.41656 13.3264V13.3266H1.17155V13.3264H3.41656ZM0.197344 14.2186H0.199219V1.78125H3.41656V1.78105H0.197344V14.2186ZM12.3861 1.58105H16V14.4186H12.3861V13.1264H14.633V2.87326H12.3861V1.58105ZM12.5834 2.67326V1.78105H15.8027V1.78125H12.5853V2.67326H12.5834ZM12.5853 13.3266V14.2186H12.5834V13.3264H14.8303V2.67345H14.8322V13.3266H12.5853ZM3.82031 5.9091C3.82031 5.18535 4.40703 4.59863 5.13078 4.59863C5.85453 4.59863 6.44124 5.18535 6.44124 5.9091C6.44124 6.56485 5.9596 7.1081 5.33078 7.2044V8.70018H5.32877C5.32982 8.75093 5.33078 8.80912 5.33078 8.87034V9.72111C5.33078 10.0195 5.57268 10.2614 5.87109 10.2614H6.24124C6.55613 10.2614 6.8114 10.5167 6.8114 10.8316C6.8114 11.1465 6.55613 11.4017 6.24124 11.4017H5.87109C4.94291 11.4017 4.19047 10.6493 4.19047 9.72111V6.82186C3.96158 6.58607 3.82031 6.26397 3.82031 5.9091ZM7.33679 5.9091C7.33679 5.59421 7.59205 5.33894 7.90694 5.33894H11.6085C11.9234 5.33894 12.1786 5.59421 12.1786 5.9091C12.1786 6.22399 11.9234 6.47925 11.6085 6.47925H7.90694C7.59205 6.47925 7.33679 6.22399 7.33679 5.9091ZM7.33679 9.86846C7.33679 9.55357 7.59205 9.2983 7.90694 9.2983H11.6085C11.9234 9.2983 12.1786 9.55357 12.1786 9.86846C12.1786 10.1833 11.9234 10.4386 11.6085 10.4386H7.90694C7.59205 10.4386 7.33679 10.1833 7.33679 9.86846Z",
        fill: "currentColor"
      }
    )
  ),
  boolean: /* @__PURE__ */ import_react.default.createElement(
    "svg",
    {
      width: "1em",
      height: "1em",
      viewBox: "0 0 16 16",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg"
    },
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        fillRule: "evenodd",
        clipRule: "evenodd",
        d: "M0 1.58105H3.6139V2.87326H1.36702V13.1264H3.6139V14.4186H0V1.58105ZM3.41656 13.3264V13.3266H1.17155V13.3264H3.41656ZM0.197344 14.2186H0.199219V1.78125H3.41656V1.78105H0.197344V14.2186ZM12.3861 1.58105H16V14.4186H12.3861V13.1264H14.633V2.87326H12.3861V1.58105ZM12.5834 2.67326V1.78105H15.8027V1.78125H12.5853V2.67326H12.5834ZM12.5853 13.3266V14.2186H12.5834V13.3264H14.8303V2.67345H14.8322V13.3266H12.5853ZM2.75 7.99993C2.75 6.14518 4.25358 4.6416 6.10833 4.6416H9.775C11.6298 4.6416 13.1333 6.14518 13.1333 7.99993C13.1333 9.85469 11.6298 11.3583 9.775 11.3583H6.10833C4.25358 11.3583 2.75 9.85469 2.75 7.99993ZM6.10833 5.85827C4.92552 5.85827 3.96667 6.81713 3.96667 7.99993C3.96667 9.18274 4.92552 10.1416 6.10833 10.1416H9.775C10.9578 10.1416 11.9167 9.18274 11.9167 7.99993C11.9167 6.81713 10.9578 5.85827 9.775 5.85827H6.10833ZM8.25 7.99993C8.25 7.1577 8.93277 6.47493 9.775 6.47493C10.6172 6.47493 11.3 7.1577 11.3 7.99993C11.3 8.84217 10.6172 9.52493 9.775 9.52493C8.93277 9.52493 8.25 8.84217 8.25 7.99993ZM9.775 7.6916C9.60471 7.6916 9.46667 7.82965 9.46667 7.99993C9.46667 8.17022 9.60471 8.30827 9.775 8.30827C9.94529 8.30827 10.0833 8.17022 10.0833 7.99993C10.0833 7.82965 9.94529 7.6916 9.775 7.6916Z",
        fill: "currentColor"
      }
    )
  ),
  string: /* @__PURE__ */ import_react.default.createElement(
    "svg",
    {
      width: "1em",
      height: "1em",
      viewBox: "0 0 16 16",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg"
    },
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        fillRule: "evenodd",
        clipRule: "evenodd",
        d: "M0 1.58105H3.6139V2.87326H1.36702V13.1264H3.6139V14.4186H0V1.58105ZM3.41656 13.3264V13.3266H1.17155V13.3264H3.41656ZM0.197344 14.2186H0.199219V1.78125H3.41656V1.78105H0.197344V14.2186ZM12.3861 1.58105H16V14.4186H12.3861V13.1264H14.633V2.87326H12.3861V1.58105ZM12.5834 2.67326V1.78105H15.8027V1.78125H12.5853V2.67326H12.5834ZM12.5853 13.3266V14.2186H12.5834V13.3264H14.8303V2.67345H14.8322V13.3266H12.5853ZM5.23701 4.07158C5.50364 3.3161 6.56205 3.28894 6.86709 4.02974L10 11.6383C10.1329 11.9609 9.979 12.3302 9.65631 12.4631C9.33363 12.596 8.96434 12.4421 8.83147 12.1194L7.8021 9.61951H4.61903L3.7474 12.0891C3.63126 12.4182 3.27034 12.5908 2.94127 12.4747C2.6122 12.3585 2.43958 11.9976 2.55573 11.6685L5.23701 4.07158ZM6.08814 5.45704L5.06505 8.35579H7.28174L6.08814 5.45704ZM8.81938 6.07534C8.81938 5.75166 9.08177 5.48926 9.40545 5.48926H12.8941C13.2178 5.48926 13.4802 5.75166 13.4802 6.07534C13.4802 6.39902 13.2178 6.66142 12.8941 6.66142H9.40545C9.08177 6.66142 8.81938 6.39902 8.81938 6.07534ZM10.2668 9.69181C10.2668 9.36812 10.5292 9.10573 10.8529 9.10573H12.8941C13.2178 9.10573 13.4802 9.36812 13.4802 9.69181C13.4802 10.0155 13.2178 10.2779 12.8941 10.2779H10.8529C10.5292 10.2779 10.2668 10.0155 10.2668 9.69181Z",
        fill: "currentColor"
      }
    )
  ),
  integer: /* @__PURE__ */ import_react.default.createElement(
    "svg",
    {
      width: "1em",
      height: "1em",
      viewBox: "0 0 16 16",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg"
    },
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        fillRule: "evenodd",
        clipRule: "evenodd",
        d: "M0 1.58105H3.6139V2.87326H1.36702V13.1264H3.6139V14.4186H0V1.58105ZM3.41656 13.3264V13.3266H1.17155V13.3264H3.41656ZM0.197344 14.2186H0.199219V1.78125H3.41656V1.78105H0.197344V14.2186ZM12.3861 1.58105H16V14.4186H12.3861V13.1264H14.633V2.87326H12.3861V1.58105ZM12.5834 2.67326V1.78105H15.8027V1.78125H12.5853V2.67326H12.5834ZM12.5853 13.3266V14.2186H12.5834V13.3264H14.8303V2.67345H14.8322V13.3266H12.5853ZM10.3614 5.22374C10.7161 4.90585 11.1581 4.75 11.6762 4.75C12.2173 4.75 12.6723 4.91467 13.0281 5.25207L13.0291 5.253C13.3852 5.59688 13.561 6.03946 13.561 6.56767C13.561 6.89 13.4945 7.17448 13.3539 7.41445C13.2572 7.57972 13.1279 7.71948 12.9685 7.83428C13.1575 7.95643 13.3099 8.11182 13.4225 8.30109C13.5793 8.5644 13.6531 8.88311 13.6531 9.24936C13.6531 9.83787 13.4612 10.3151 13.0656 10.6612C12.6982 10.9795 12.2305 11.1341 11.6762 11.1341C11.1356 11.1341 10.6805 10.9925 10.324 10.6977C9.92124 10.3691 9.71723 9.90026 9.69942 9.31256L9.69473 9.15802H10.846L10.8539 9.2997C10.8689 9.5698 10.9591 9.75553 11.1096 9.87941L11.1106 9.88027C11.2519 9.99882 11.4365 10.0631 11.6762 10.0631C11.9765 10.0631 12.1743 9.98692 12.2984 9.86071C12.4229 9.73404 12.4984 9.53136 12.4984 9.22422C12.4984 8.92116 12.4215 8.72127 12.2939 8.59581C12.1658 8.46989 11.961 8.39373 11.6511 8.39373H11.3586V7.34788H11.6511C11.9297 7.34788 12.111 7.27834 12.2238 7.16555C12.3366 7.05276 12.4062 6.87138 12.4062 6.59281C12.4062 6.30696 12.3378 6.12041 12.2277 6.00501C12.1188 5.89092 11.9446 5.82098 11.6762 5.82098C11.4248 5.82098 11.2539 5.88537 11.1407 5.99325C11.0268 6.10185 10.9497 6.27522 10.9291 6.5375L10.9183 6.67577H9.76788L9.77492 6.51904C9.79886 5.98644 9.99237 5.54989 10.3614 5.22374ZM5.91032 5.26037C6.26612 4.92297 6.72112 4.7583 7.26219 4.7583C7.80751 4.7583 8.26297 4.91938 8.61401 5.25194L8.61501 5.25289C8.96719 5.59272 9.13852 6.04185 9.13852 6.58435C9.13852 6.84997 9.08709 7.09565 8.9817 7.31883L8.98114 7.31999C8.89563 7.49712 8.74775 7.71415 8.54418 7.96862L8.54322 7.96981L6.87446 10.0127H9.13852V11.0753H5.36909V10.1089L7.69946 7.27679C7.89456 7.04062 7.98374 6.80773 7.98374 6.57597C7.98374 6.29602 7.91626 6.11385 7.8078 6.00122C7.70036 5.88964 7.52811 5.8209 7.26219 5.8209C7.04017 5.8209 6.87439 5.88173 6.75075 5.99193C6.61227 6.11766 6.53226 6.30918 6.53226 6.59273V6.74273H5.37747V6.59273C5.37747 6.05443 5.55248 5.60586 5.90934 5.2613L5.91032 5.26037ZM3.50907 4.80865H4.56964V11.0754H3.41486V6.2201L2.25 7.24249V5.89561L3.50907 4.80865Z",
        fill: "currentColor"
      }
    )
  ),
  number: /* @__PURE__ */ import_react.default.createElement(
    "svg",
    {
      width: "1em",
      height: "1em",
      viewBox: "0 0 16 16",
      fill: "none",
      xmlns: "http://www.w3.org/2000/svg"
    },
    /* @__PURE__ */ import_react.default.createElement(
      "path",
      {
        fillRule: "evenodd",
        clipRule: "evenodd",
        d: "M3.6139 1.58154H0V14.4191H3.6139V13.1269H1.36702V2.87375H3.6139V1.58154ZM3.41656 13.3271V13.3269H1.17155V13.3271H3.41656ZM0.199219 14.2191H0.197344V1.78154H3.41656V1.78174H0.199219V14.2191ZM16 1.58154H12.3861V2.87375H14.633V13.1269H12.3861V14.4191H16V1.58154ZM12.5834 1.78154V2.67375H12.5853V1.78174H15.8027V1.78154H12.5834ZM12.5853 14.2191V13.3271H14.8322V2.67394H14.8303V13.3269H12.5834V14.2191H12.5853ZM6.86771 4.5C5.87019 4.5 5.00104 5.30767 5.00104 6.31667V9.63333C5.00104 10.6423 5.87019 11.45 6.86771 11.45C7.86523 11.45 8.73438 10.6423 8.73438 9.63333V6.31667C8.73438 5.30767 7.86523 4.5 6.86771 4.5ZM11.1177 4.5C10.1202 4.5 9.25104 5.30767 9.25104 6.31667V9.63333C9.25104 10.6423 10.1202 11.45 11.1177 11.45C12.1152 11.45 12.9844 10.6423 12.9844 9.63333V6.31667C12.9844 5.30767 12.1152 4.5 11.1177 4.5ZM6.13438 6.31667C6.13438 5.9503 6.47884 5.63333 6.86771 5.63333C7.25657 5.63333 7.60104 5.9503 7.60104 6.31667V9.63333C7.60104 9.9997 7.25657 10.3167 6.86771 10.3167C6.47884 10.3167 6.13438 9.9997 6.13438 9.63333V6.31667ZM10.3844 6.31667C10.3844 5.9503 10.7288 5.63333 11.1177 5.63333C11.5066 5.63333 11.851 5.9503 11.851 6.31667V9.63333C11.851 9.9997 11.5066 10.3167 11.1177 10.3167C10.7288 10.3167 10.3844 9.9997 10.3844 9.63333V6.31667ZM3.75938 9.85C3.33135 9.85 2.98438 10.197 2.98438 10.625C2.98438 11.053 3.33135 11.4 3.75938 11.4C4.1874 11.4 4.53438 11.053 4.53438 10.625C4.53438 10.197 4.1874 9.85 3.75938 9.85Z",
        fill: "currentColor"
      }
    )
  )
};
var getSchemaIcon = (value) => {
  if (value?.type === "array") {
    return ArrayIcons[value.items?.type || "object"];
  }
  return VariableTypeIcons[value?.type || "object"];
};
var labelStyle = { display: "flex", alignItems: "center", gap: 5 };
var firstUppercase = (str) => str.charAt(0).toUpperCase() + str.slice(1);
var baseOptions = [
  {
    label: /* @__PURE__ */ import_react.default.createElement("div", { style: labelStyle }, /* @__PURE__ */ import_react.default.createElement(import_semi_icons.default, { size: "small", svg: getSchemaIcon({ type: "string" }) }), firstUppercase("string")),
    value: "string"
  },
  {
    label: /* @__PURE__ */ import_react.default.createElement("div", { style: labelStyle }, /* @__PURE__ */ import_react.default.createElement(import_semi_icons.default, { size: "small", svg: getSchemaIcon({ type: "integer" }) }), firstUppercase("integer")),
    value: "integer"
  },
  {
    label: /* @__PURE__ */ import_react.default.createElement("div", { style: labelStyle }, /* @__PURE__ */ import_react.default.createElement(import_semi_icons.default, { size: "small", svg: getSchemaIcon({ type: "number" }) }), firstUppercase("number")),
    value: "number"
  },
  {
    label: /* @__PURE__ */ import_react.default.createElement("div", { style: labelStyle }, /* @__PURE__ */ import_react.default.createElement(import_semi_icons.default, { size: "small", svg: getSchemaIcon({ type: "boolean" }) }), firstUppercase("boolean")),
    value: "boolean"
  },
  {
    label: /* @__PURE__ */ import_react.default.createElement("div", { style: labelStyle }, /* @__PURE__ */ import_react.default.createElement(import_semi_icons.default, { size: "small", svg: getSchemaIcon({ type: "object" }) }), firstUppercase("object")),
    value: "object"
  }
];
var options = [
  ...baseOptions,
  {
    label: /* @__PURE__ */ import_react.default.createElement("div", { style: labelStyle }, /* @__PURE__ */ import_react.default.createElement(import_semi_icons.default, { size: "small", svg: getSchemaIcon({ type: "array" }) }), firstUppercase("array")),
    value: "array",
    children: baseOptions.map((_opt) => ({
      ..._opt,
      value: `${_opt.value}`,
      label: /* @__PURE__ */ import_react.default.createElement("div", { style: labelStyle }, /* @__PURE__ */ import_react.default.createElement(
        import_semi_icons.default,
        {
          size: "small",
          svg: getSchemaIcon({ type: "array", items: { type: _opt.value } })
        }
      ), firstUppercase(_opt.value))
    }))
  }
];

// src/utils/json-schema/index.ts
var import_lodash = require("lodash");
var import_editor = require("@flowgram.ai/editor");
var JsonSchemaUtils;
((JsonSchemaUtils2) => {
  function schemaToAST(jsonSchema) {
    const { type, extra } = jsonSchema || {};
    const { weak = false } = extra || {};
    if (!type) {
      return void 0;
    }
    switch (type) {
      case "object":
        if (weak) {
          return { kind: import_editor.ASTKind.Object, weak: true };
        }
        return import_editor.ASTFactory.createObject({
          properties: Object.entries(jsonSchema.properties || {}).sort((a, b) => ((0, import_lodash.get)(a?.[1], "extra.index") || 0) - ((0, import_lodash.get)(b?.[1], "extra.index") || 0)).map(([key, _property]) => ({
            key,
            type: schemaToAST(_property),
            meta: { description: _property.description }
          }))
        });
      case "array":
        if (weak) {
          return { kind: import_editor.ASTKind.Array, weak: true };
        }
        return import_editor.ASTFactory.createArray({
          items: schemaToAST(jsonSchema.items)
        });
      case "map":
        if (weak) {
          return { kind: import_editor.ASTKind.Map, weak: true };
        }
        return import_editor.ASTFactory.createMap({
          valueType: schemaToAST(jsonSchema.additionalProperties)
        });
      case "string":
        return import_editor.ASTFactory.createString();
      case "number":
        return import_editor.ASTFactory.createNumber();
      case "boolean":
        return import_editor.ASTFactory.createBoolean();
      case "integer":
        return import_editor.ASTFactory.createInteger();
      default:
        return import_editor.ASTFactory.createCustomType({ typeName: type });
    }
  }
  JsonSchemaUtils2.schemaToAST = schemaToAST;
  function astToSchema(typeAST, options2) {
    const { drilldown = true } = options2 || {};
    if (import_editor.ASTMatch.isString(typeAST)) {
      return {
        type: "string"
      };
    }
    if (import_editor.ASTMatch.isBoolean(typeAST)) {
      return {
        type: "boolean"
      };
    }
    if (import_editor.ASTMatch.isNumber(typeAST)) {
      return {
        type: "number"
      };
    }
    if (import_editor.ASTMatch.isInteger(typeAST)) {
      return {
        type: "integer"
      };
    }
    if (import_editor.ASTMatch.isObject(typeAST)) {
      return {
        type: "object",
        properties: drilldown ? Object.fromEntries(
          Object.entries(typeAST.properties).map(([key, value]) => [key, astToSchema(value)])
        ) : {}
      };
    }
    if (import_editor.ASTMatch.isArray(typeAST)) {
      return {
        type: "array",
        items: drilldown ? astToSchema(typeAST.items) : void 0
      };
    }
    if (import_editor.ASTMatch.isMap(typeAST)) {
      return {
        type: "map",
        items: drilldown ? astToSchema(typeAST.valueType) : void 0
      };
    }
    if (import_editor.ASTMatch.isCustomType(typeAST)) {
      return {
        type: typeAST.typeName
      };
    }
    return void 0;
  }
  JsonSchemaUtils2.astToSchema = astToSchema;
  function isASTMatchSchema(typeAST, schema) {
    if (Array.isArray(schema)) {
      return typeAST.isTypeEqual(
        import_editor.ASTFactory.createUnion({
          types: schema.map((_schema) => schemaToAST(_schema)).filter(Boolean)
        })
      );
    }
    return typeAST.isTypeEqual(schemaToAST(schema));
  }
  JsonSchemaUtils2.isASTMatchSchema = isASTMatchSchema;
})(JsonSchemaUtils || (JsonSchemaUtils = {}));

// src/components/variable-selector/use-variable-tree.tsx
function useVariableTree(params) {
  const { includeSchema, excludeSchema } = params;
  const available = (0, import_editor2.useScopeAvailable)();
  const getVariableTypeIcon = (0, import_react2.useCallback)((variable) => {
    if (variable.meta?.icon) {
      if (typeof variable.meta.icon === "string") {
        return /* @__PURE__ */ import_react2.default.createElement("img", { style: { marginRight: 8 }, width: 12, height: 12, src: variable.meta.icon });
      }
      return variable.meta.icon;
    }
    const _type = variable.type;
    if (import_editor2.ASTMatch.isArray(_type)) {
      return /* @__PURE__ */ import_react2.default.createElement(
        import_semi_ui.Icon,
        {
          size: "small",
          svg: ArrayIcons[_type.items?.kind.toLowerCase()] || VariableTypeIcons.array
        }
      );
    }
    if (import_editor2.ASTMatch.isCustomType(_type)) {
      return /* @__PURE__ */ import_react2.default.createElement(import_semi_ui.Icon, { size: "small", svg: VariableTypeIcons[_type.typeName.toLowerCase()] });
    }
    return /* @__PURE__ */ import_react2.default.createElement(import_semi_ui.Icon, { size: "small", svg: VariableTypeIcons[variable.type?.kind.toLowerCase()] });
  }, []);
  const renderVariable = (variable, parentFields = []) => {
    let type = variable?.type;
    if (!type) {
      return null;
    }
    let children;
    if (import_editor2.ASTMatch.isObject(type)) {
      children = (type.properties || []).map((_property) => renderVariable(_property, [...parentFields, variable])).filter(Boolean);
      if (!children?.length) {
        return null;
      }
    }
    const keyPath = [...parentFields.map((_field) => _field.key), variable.key];
    const key = keyPath.join(".");
    const isSchemaInclude = includeSchema ? JsonSchemaUtils.isASTMatchSchema(type, includeSchema) : true;
    const isSchemaExclude = excludeSchema ? JsonSchemaUtils.isASTMatchSchema(type, excludeSchema) : false;
    const isSchemaMatch = isSchemaInclude && !isSchemaExclude;
    if (!isSchemaMatch && !children?.length) {
      return null;
    }
    return {
      key,
      label: variable.meta?.title || variable.key,
      value: key,
      keyPath,
      icon: getVariableTypeIcon(variable),
      children,
      disabled: !isSchemaMatch,
      rootMeta: parentFields[0]?.meta
    };
  };
  return [...available.variables.slice(0).reverse()].map((_variable) => renderVariable(_variable)).filter(Boolean);
}

// src/components/variable-selector/styles.tsx
var import_styled_components = __toESM(require("styled-components"), 1);
var import_semi_ui2 = require("@douyinfe/semi-ui");
var UIRootTitle = import_styled_components.default.span`
  margin-right: 4px;
  color: var(--semi-color-text-2);
`;
var UITag = (0, import_styled_components.default)(import_semi_ui2.Tag)`
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: flex-start;

  & .semi-tag-content-center {
    justify-content: flex-start;
  }

  &.semi-tag {
    margin: 0;
  }
`;
var UITreeSelect = (0, import_styled_components.default)(import_semi_ui2.TreeSelect)`
  outline: ${({ $error }) => $error ? "1px solid red" : "none"};

  height: 22px;
  min-height: 22px;
  line-height: 22px;

  & .semi-tree-select-selection {
    padding: 0 2px;
    height: 22px;
  }

  & .semi-tree-select-selection-content {
    width: 100%;
  }

  & .semi-tree-select-selection-placeholder {
    padding-left: 10px;
  }
`;

// src/components/variable-selector/index.tsx
var VariableSelector = ({
  value,
  config,
  onChange,
  style,
  readonly = false,
  includeSchema,
  excludeSchema,
  hasError,
  triggerRender
}) => {
  const treeData = useVariableTree({ includeSchema, excludeSchema });
  const treeValue = (0, import_react3.useMemo)(() => {
    if (typeof value === "string") {
      console.warn(
        "The Value of VariableSelector is a string, it should be an ARRAY. \n",
        "Please check the value of VariableSelector \n"
      );
      return value;
    }
    return value?.join(".");
  }, [value]);
  const renderIcon = (icon) => {
    if (typeof icon === "string") {
      return /* @__PURE__ */ import_react3.default.createElement("img", { style: { marginRight: 8 }, width: 12, height: 12, src: icon });
    }
    return icon;
  };
  return /* @__PURE__ */ import_react3.default.createElement(import_react3.default.Fragment, null, /* @__PURE__ */ import_react3.default.createElement(
    UITreeSelect,
    {
      dropdownMatchSelectWidth: false,
      disabled: readonly,
      treeData,
      size: "small",
      value: treeValue,
      clearIcon: null,
      $error: hasError,
      style,
      validateStatus: hasError ? "error" : void 0,
      onChange: (_, _config) => {
        onChange(_config.keyPath);
      },
      renderSelectedItem: (_option) => {
        if (!_option?.keyPath) {
          return /* @__PURE__ */ import_react3.default.createElement(
            UITag,
            {
              prefixIcon: /* @__PURE__ */ import_react3.default.createElement(import_semi_icons2.IconIssueStroked, null),
              color: "amber",
              closable: !readonly,
              onClose: () => onChange(void 0)
            },
            config?.notFoundContent ?? "Undefined"
          );
        }
        return /* @__PURE__ */ import_react3.default.createElement(
          UITag,
          {
            prefixIcon: renderIcon(_option.rootMeta?.icon || _option?.icon),
            closable: !readonly,
            onClose: () => onChange(void 0)
          },
          /* @__PURE__ */ import_react3.default.createElement(UIRootTitle, null, _option.rootMeta?.title ? `${_option.rootMeta?.title} -` : null),
          _option.label
        );
      },
      showClear: false,
      arrowIcon: /* @__PURE__ */ import_react3.default.createElement(import_semi_icons2.IconChevronDownStroked, { size: "small" }),
      triggerRender,
      placeholder: config?.placeholder ?? "Select Variable..."
    }
  ));
};

// src/components/type-selector/index.tsx
var import_react4 = __toESM(require("react"), 1);
var import_semi_ui3 = require("@douyinfe/semi-ui");
var getTypeSelectValue = (value) => {
  if (value?.type === "array" && value?.items) {
    return [value.type, ...getTypeSelectValue(value.items) || []];
  }
  return value?.type ? [value.type] : void 0;
};
var parseTypeSelectValue = (value) => {
  const [type, ...subTypes] = value || [];
  if (type === "array") {
    return { type: "array", items: parseTypeSelectValue(subTypes) };
  }
  return { type };
};
function TypeSelector(props) {
  const { value, onChange, disabled, style } = props;
  const selectValue = (0, import_react4.useMemo)(() => getTypeSelectValue(value), [value]);
  return /* @__PURE__ */ import_react4.default.createElement(
    import_semi_ui3.Cascader,
    {
      disabled,
      size: "small",
      triggerRender: () => /* @__PURE__ */ import_react4.default.createElement(import_semi_ui3.Button, { size: "small", style }, getSchemaIcon(value)),
      treeData: options,
      value: selectValue,
      leafOnly: true,
      onChange: (value2) => {
        onChange(parseTypeSelectValue(value2));
      }
    }
  );
}

// src/components/json-schema-editor/index.tsx
var import_react8 = __toESM(require("react"), 1);
var import_semi_ui4 = require("@douyinfe/semi-ui");
var import_semi_icons4 = require("@douyinfe/semi-icons");

// src/components/json-schema-editor/styles.tsx
var import_react5 = __toESM(require("react"), 1);
var import_styled_components2 = __toESM(require("styled-components"), 1);
var import_semi_icons3 = __toESM(require("@douyinfe/semi-icons"), 1);
var UIContainer = import_styled_components2.default.div`
  /* & .semi-input {
    background-color: #fff;
    border-radius: 6px;
    height: 24px;
  } */
`;
var UIRow = import_styled_components2.default.div`
  display: flex;
  align-items: center;
  gap: 6px;
`;
var UICollapseTrigger = import_styled_components2.default.div`
  cursor: pointer;
  margin-right: 5px;
`;
var UIExpandDetail = import_styled_components2.default.div`
  display: flex;
  flex-direction: column;
`;
var UILabel = import_styled_components2.default.div`
  font-size: 12px;
  color: #999;
  font-weight: 400;
  margin-bottom: 2px;
`;
var UIProperties = import_styled_components2.default.div`
  display: grid;
  grid-template-columns: auto 1fr;

  ${({ $shrink }) => $shrink && import_styled_components2.css`
      padding-left: 10px;
      margin-top: 10px;
    `}
`;
var UIPropertyLeft = import_styled_components2.default.div`
  grid-column: 1;
  position: relative;

  ${({ $showLine, $isLast }) => $showLine && import_styled_components2.css`
      &::before {
        /* 竖线 */
        content: '';
        position: absolute;
        left: -22px;
        top: -18px;
        bottom: ${$isLast ? "12px" : "0px"};
        width: 1px;
        background: #d9d9d9;
        display: block;
      }

      &::after {
        /* 横线 */
        content: '';
        position: absolute;
        left: -22px; // 横线起点和竖线对齐
        top: 12px; // 跟随你的行高调整
        width: 22px; // 横线长度
        height: 1px;
        background: #d9d9d9;
        display: block;
      }
    `}
`;
var UIPropertyRight = import_styled_components2.default.div`
  grid-column: 2;
  margin-bottom: 10px;

  &:last-child {
    margin-bottom: 0px;
  }
`;
var UIPropertyMain = import_styled_components2.default.div`
  display: flex;
  flex-direction: column;
  gap: 10px;

  ${({ $expand }) => $expand && import_styled_components2.css`
      background-color: #f5f5f5;
      padding: 10px;
      border-radius: 4px;
    `}
`;
var UICollapsible = import_styled_components2.default.div`
  display: none;

  ${({ $collapse }) => $collapse && import_styled_components2.css`
      display: block;
    `}
`;
var UIName = import_styled_components2.default.div`
  flex-grow: 1;
`;
var UIType = import_styled_components2.default.div``;
var UIRequired = import_styled_components2.default.div``;
var UIActions = import_styled_components2.default.div`
  white-space: nowrap;
`;
var iconAddChildrenSvg = /* @__PURE__ */ import_react5.default.createElement(
  "svg",
  {
    className: "icon-icon icon-icon-coz_add_node ",
    width: "1em",
    height: "1em",
    viewBox: "0 0 24 24",
    fill: "currentColor",
    xmlns: "http://www.w3.org/2000/svg"
  },
  /* @__PURE__ */ import_react5.default.createElement(
    "path",
    {
      fillRule: "evenodd",
      clipRule: "evenodd",
      d: "M11 6.49988C11 8.64148 9.50397 10.4337 7.49995 10.8884V15.4998C7.49995 16.0521 7.94767 16.4998 8.49995 16.4998H11.208C11.0742 16.8061 11 17.1443 11 17.4998C11 17.8554 11.0742 18.1936 11.208 18.4998H8.49995C6.8431 18.4998 5.49995 17.1567 5.49995 15.4998V10.8884C3.49599 10.4336 2 8.64145 2 6.49988C2 4.0146 4.01472 1.99988 6.5 1.99988C8.98528 1.99988 11 4.0146 11 6.49988ZM6.5 8.99988C7.88071 8.99988 9 7.88059 9 6.49988C9 5.11917 7.88071 3.99988 6.5 3.99988C5.11929 3.99988 4 5.11917 4 6.49988C4 7.88059 5.11929 8.99988 6.5 8.99988Z"
    }
  ),
  /* @__PURE__ */ import_react5.default.createElement("path", { d: "M17.5 12.4999C18.0523 12.4999 18.5 12.9476 18.5 13.4999V16.4999H21.5C22.0523 16.4999 22.5 16.9476 22.5 17.4999C22.5 18.0522 22.0523 18.4999 21.5 18.4999H18.5V21.4999C18.5 22.0522 18.0523 22.4999 17.5 22.4999C16.9477 22.4999 16.5 22.0522 16.5 21.4999V18.4999H13.5C12.9477 18.4999 12.5 18.0522 12.5 17.4999C12.5 16.9476 12.9477 16.4999 13.5 16.4999H16.5V13.4999C16.5 12.9476 16.9477 12.4999 17.5 12.4999Z" })
);
var IconAddChildren = () => /* @__PURE__ */ import_react5.default.createElement(import_semi_icons3.default, { size: "small", svg: iconAddChildrenSvg });

// src/components/json-schema-editor/hooks.tsx
var import_react6 = require("react");
var _id = 0;
function genId() {
  return _id++;
}
function getDrilldownSchema(value, path) {
  if (!value) {
    return {};
  }
  if (value.type === "array" && value.items) {
    return getDrilldownSchema(value.items, [...path || [], "items"]);
  }
  return { schema: value, path };
}
function usePropertiesEdit(value, onChange) {
  const drilldown = (0, import_react6.useMemo)(() => getDrilldownSchema(value), [value, value?.type, value?.items]);
  const isDrilldownObject = drilldown.schema?.type === "object";
  const initPropertyList = (0, import_react6.useMemo)(
    () => isDrilldownObject ? Object.entries(drilldown.schema?.properties || {}).sort(([, a], [, b]) => (a.extra?.index ?? 0) - (b.extra?.index ?? 0)).map(
      ([name, _value], index) => ({
        key: genId(),
        name,
        isPropertyRequired: drilldown.schema?.required?.includes(name) || false,
        ..._value,
        extra: {
          ..._value.extra || {},
          index
        }
      })
    ) : [],
    [isDrilldownObject]
  );
  const [propertyList, setPropertyList] = (0, import_react6.useState)(initPropertyList);
  const mountRef = (0, import_react6.useRef)(false);
  (0, import_react6.useEffect)(() => {
    if (mountRef.current) {
      setPropertyList((_list) => {
        const nameMap = /* @__PURE__ */ new Map();
        for (const _property of _list) {
          if (_property.name) {
            nameMap.set(_property.name, _property);
          }
        }
        return Object.entries(drilldown.schema?.properties || {}).sort(([, a], [, b]) => (a.extra?.index ?? 0) - (b.extra?.index ?? 0)).map(([name, _value]) => {
          const _property = nameMap.get(name);
          if (_property) {
            return {
              key: _property.key,
              name,
              isPropertyRequired: drilldown.schema?.required?.includes(name) || false,
              ..._value
            };
          }
          return {
            key: genId(),
            name,
            isPropertyRequired: drilldown.schema?.required?.includes(name) || false,
            ..._value
          };
        });
      });
    }
    mountRef.current = true;
  }, [drilldown.schema]);
  const updatePropertyList = (updater) => {
    setPropertyList((_list) => {
      const next = updater(_list);
      const nextProperties = {};
      const nextRequired = [];
      for (const _property of next) {
        if (!_property.name) {
          continue;
        }
        nextProperties[_property.name] = _property;
        if (_property.isPropertyRequired) {
          nextRequired.push(_property.name);
        }
      }
      let drilldownSchema = value || {};
      if (drilldown.path) {
        drilldownSchema = drilldown.path.reduce((acc, key) => acc[key], value || {});
      }
      drilldownSchema.properties = nextProperties;
      drilldownSchema.required = nextRequired;
      onChange?.(value || {});
      return next;
    });
  };
  const onAddProperty = () => {
    updatePropertyList((_list) => [
      ..._list,
      { key: genId(), name: "", type: "string", extra: { index: _list.length + 1 } }
    ]);
  };
  const onRemoveProperty = (key) => {
    updatePropertyList((_list) => _list.filter((_property) => _property.key !== key));
  };
  const onEditProperty = (key, nextValue) => {
    updatePropertyList(
      (_list) => _list.map((_property) => _property.key === key ? nextValue : _property)
    );
  };
  (0, import_react6.useEffect)(() => {
    if (!isDrilldownObject) {
      setPropertyList([]);
    }
  }, [isDrilldownObject]);
  return {
    propertyList,
    isDrilldownObject,
    onAddProperty,
    onRemoveProperty,
    onEditProperty
  };
}

// src/components/json-schema-editor/components/blur-input.tsx
var import_react7 = __toESM(require("react"), 1);
var import_input = __toESM(require("@douyinfe/semi-ui/lib/es/input"), 1);
function BlurInput(props) {
  const [value, setValue] = (0, import_react7.useState)("");
  (0, import_react7.useEffect)(() => {
    setValue(props.value);
  }, [props.value]);
  return /* @__PURE__ */ import_react7.default.createElement(
    import_input.default,
    {
      ...props,
      value,
      onChange: (value2) => {
        setValue(value2);
      },
      onBlur: (e) => props.onChange?.(value, e)
    }
  );
}

// src/components/json-schema-editor/index.tsx
function JsonSchemaEditor(props) {
  const { value = { type: "object" }, config = {}, onChange: onChangeProps } = props;
  const { propertyList, onAddProperty, onRemoveProperty, onEditProperty } = usePropertiesEdit(
    value,
    onChangeProps
  );
  return /* @__PURE__ */ import_react8.default.createElement(UIContainer, null, /* @__PURE__ */ import_react8.default.createElement(UIProperties, null, propertyList.map((_property) => /* @__PURE__ */ import_react8.default.createElement(
    PropertyEdit,
    {
      key: _property.key,
      value: _property,
      config,
      onChange: (_v) => {
        onEditProperty(_property.key, _v);
      },
      onRemove: () => {
        onRemoveProperty(_property.key);
      }
    }
  ))), /* @__PURE__ */ import_react8.default.createElement(import_semi_ui4.Button, { size: "small", style: { marginTop: 10 }, icon: /* @__PURE__ */ import_react8.default.createElement(import_semi_icons4.IconPlus, null), onClick: onAddProperty }, config?.addButtonText ?? "Add"));
}
function PropertyEdit(props) {
  const { value, config, onChange: onChangeProps, onRemove, $isLast, $showLine } = props;
  const [expand, setExpand] = (0, import_react8.useState)(false);
  const [collapse, setCollapse] = (0, import_react8.useState)(false);
  const { name, type, items, description, isPropertyRequired } = value || {};
  const typeSelectorValue = (0, import_react8.useMemo)(() => ({ type, items }), [type, items]);
  const { propertyList, isDrilldownObject, onAddProperty, onRemoveProperty, onEditProperty } = usePropertiesEdit(value, onChangeProps);
  const onChange = (key, _value) => {
    onChangeProps?.({
      ...value || {},
      [key]: _value
    });
  };
  const showCollapse = isDrilldownObject && propertyList.length > 0;
  return /* @__PURE__ */ import_react8.default.createElement(import_react8.default.Fragment, null, /* @__PURE__ */ import_react8.default.createElement(UIPropertyLeft, { $isLast, $showLine }, showCollapse && /* @__PURE__ */ import_react8.default.createElement(UICollapseTrigger, { onClick: () => setCollapse((_collapse) => !_collapse) }, collapse ? /* @__PURE__ */ import_react8.default.createElement(import_semi_icons4.IconChevronDown, { size: "small" }) : /* @__PURE__ */ import_react8.default.createElement(import_semi_icons4.IconChevronRight, { size: "small" }))), /* @__PURE__ */ import_react8.default.createElement(UIPropertyRight, null, /* @__PURE__ */ import_react8.default.createElement(UIPropertyMain, { $expand: expand }, /* @__PURE__ */ import_react8.default.createElement(UIRow, null, /* @__PURE__ */ import_react8.default.createElement(UIName, null, /* @__PURE__ */ import_react8.default.createElement(
    BlurInput,
    {
      placeholder: config?.placeholder ?? "Input Variable Name",
      size: "small",
      value: name,
      onChange: (value2) => onChange("name", value2)
    }
  )), /* @__PURE__ */ import_react8.default.createElement(UIType, null, /* @__PURE__ */ import_react8.default.createElement(
    TypeSelector,
    {
      value: typeSelectorValue,
      onChange: (_value) => {
        onChangeProps?.({
          ...value || {},
          ..._value
        });
      }
    }
  )), /* @__PURE__ */ import_react8.default.createElement(UIRequired, null, /* @__PURE__ */ import_react8.default.createElement(
    import_semi_ui4.Checkbox,
    {
      checked: isPropertyRequired,
      onChange: (e) => onChange("isPropertyRequired", e.target.checked)
    }
  )), /* @__PURE__ */ import_react8.default.createElement(UIActions, null, /* @__PURE__ */ import_react8.default.createElement(
    import_semi_ui4.IconButton,
    {
      size: "small",
      theme: "borderless",
      icon: expand ? /* @__PURE__ */ import_react8.default.createElement(import_semi_icons4.IconShrink, { size: "small" }) : /* @__PURE__ */ import_react8.default.createElement(import_semi_icons4.IconExpand, { size: "small" }),
      onClick: () => setExpand((_expand) => !_expand)
    }
  ), isDrilldownObject && /* @__PURE__ */ import_react8.default.createElement(
    import_semi_ui4.IconButton,
    {
      size: "small",
      theme: "borderless",
      icon: /* @__PURE__ */ import_react8.default.createElement(IconAddChildren, null),
      onClick: () => {
        onAddProperty();
        setCollapse(true);
      }
    }
  ), /* @__PURE__ */ import_react8.default.createElement(
    import_semi_ui4.IconButton,
    {
      size: "small",
      theme: "borderless",
      icon: /* @__PURE__ */ import_react8.default.createElement(import_semi_icons4.IconMinus, { size: "small" }),
      onClick: onRemove
    }
  ))), expand && /* @__PURE__ */ import_react8.default.createElement(UIExpandDetail, null, /* @__PURE__ */ import_react8.default.createElement(UILabel, null, config?.descTitle ?? "Description"), /* @__PURE__ */ import_react8.default.createElement(
    BlurInput,
    {
      size: "small",
      value: description,
      onChange: (value2) => onChange("description", value2),
      placeholder: config?.descPlaceholder ?? "Help LLM to understand the property"
    }
  ))), showCollapse && /* @__PURE__ */ import_react8.default.createElement(UICollapsible, { $collapse: collapse }, /* @__PURE__ */ import_react8.default.createElement(UIProperties, { $shrink: true }, propertyList.map((_property, index) => /* @__PURE__ */ import_react8.default.createElement(
    PropertyEdit,
    {
      key: _property.key,
      value: _property,
      config,
      onChange: (_v) => {
        onEditProperty(_property.key, _v);
      },
      onRemove: () => {
        onRemoveProperty(_property.key);
      },
      $isLast: index === propertyList.length - 1,
      $showLine: true
    }
  ))))));
}

// src/components/batch-variable-selector/index.tsx
var import_react9 = __toESM(require("react"), 1);
var import_editor3 = require("@flowgram.ai/editor");
var batchVariableSchema = {
  type: "array",
  extra: { weak: true }
};
function BatchVariableSelector(props) {
  return /* @__PURE__ */ import_react9.default.createElement(import_editor3.PrivateScopeProvider, null, /* @__PURE__ */ import_react9.default.createElement(VariableSelector, { ...props, includeSchema: batchVariableSchema }));
}

// src/components/constant-input/index.tsx
var import_react10 = __toESM(require("react"), 1);
var import_semi_ui5 = require("@douyinfe/semi-ui");
var defaultStrategies = [
  {
    hit: (schema) => schema?.type === "string",
    Renderer: (props) => /* @__PURE__ */ import_react10.default.createElement(import_semi_ui5.Input, { placeholder: "Please Input String", size: "small", disabled: props.readonly, ...props })
  },
  {
    hit: (schema) => schema?.type === "number",
    Renderer: (props) => /* @__PURE__ */ import_react10.default.createElement(
      import_semi_ui5.InputNumber,
      {
        placeholder: "Please Input Number",
        size: "small",
        disabled: props.readonly,
        hideButtons: true,
        ...props
      }
    )
  },
  {
    hit: (schema) => schema?.type === "integer",
    Renderer: (props) => /* @__PURE__ */ import_react10.default.createElement(
      import_semi_ui5.InputNumber,
      {
        placeholder: "Please Input Integer",
        size: "small",
        disabled: props.readonly,
        hideButtons: true,
        precision: 0,
        ...props
      }
    )
  },
  {
    hit: (schema) => schema?.type === "boolean",
    Renderer: (props) => {
      const { value, onChange, ...rest } = props;
      return /* @__PURE__ */ import_react10.default.createElement(
        import_semi_ui5.Select,
        {
          placeholder: "Please Select Boolean",
          size: "small",
          disabled: props.readonly,
          optionList: [
            { label: "True", value: 1 },
            { label: "False", value: 0 }
          ],
          value: value ? 1 : 0,
          onChange: (value2) => onChange?.(!!value2),
          ...rest
        }
      );
    }
  }
];
function ConstantInput(props) {
  const { value, onChange, schema, strategies: extraStrategies, readonly, ...rest } = props;
  const strategies = (0, import_react10.useMemo)(
    () => [...defaultStrategies, ...extraStrategies || []],
    [extraStrategies]
  );
  const Renderer = (0, import_react10.useMemo)(() => {
    const strategy = strategies.find((_strategy) => _strategy.hit(schema));
    return strategy?.Renderer;
  }, [strategies, schema]);
  if (!Renderer) {
    return /* @__PURE__ */ import_react10.default.createElement(import_semi_ui5.Input, { size: "small", disabled: true, placeholder: "Unsupported type" });
  }
  return /* @__PURE__ */ import_react10.default.createElement(Renderer, { value, onChange, readonly, ...rest });
}

// src/components/dynamic-value-input/index.tsx
var import_react11 = __toESM(require("react"), 1);
var import_semi_ui6 = require("@douyinfe/semi-ui");
var import_semi_icons5 = require("@douyinfe/semi-icons");

// src/components/dynamic-value-input/styles.tsx
var import_styled_components3 = __toESM(require("styled-components"), 1);
var UIContainer2 = import_styled_components3.default.div`
  display: flex;
  align-items: center;
  gap: 5px;
`;
var UIMain = import_styled_components3.default.div`
  flex-grow: 1;

  & .semi-tree-select,
  & .semi-input-number,
  & .semi-select {
    width: 100%;
  }
`;
var UITrigger = import_styled_components3.default.div``;

// src/components/dynamic-value-input/index.tsx
function DynamicValueInput({
  value,
  onChange,
  readonly,
  style,
  schema,
  constantProps
}) {
  const includeSchema = (0, import_react11.useMemo)(() => {
    if (schema?.type === "number") {
      return [schema, { type: "integer" }];
    }
    return schema;
  }, [schema]);
  const renderMain = () => {
    if (value?.type === "ref") {
      return /* @__PURE__ */ import_react11.default.createElement(
        VariableSelector,
        {
          value: value?.content,
          onChange: (_v) => onChange(_v ? { type: "ref", content: _v } : void 0),
          includeSchema,
          readonly
        }
      );
    }
    return /* @__PURE__ */ import_react11.default.createElement(
      ConstantInput,
      {
        value: value?.content,
        onChange: (_v) => onChange({ type: "constant", content: _v }),
        schema: schema || { type: "string" },
        readonly,
        ...constantProps
      }
    );
  };
  const renderTrigger = () => /* @__PURE__ */ import_react11.default.createElement(
    VariableSelector,
    {
      style: { width: "100%" },
      value: value?.type === "ref" ? value?.content : void 0,
      onChange: (_v) => onChange({ type: "ref", content: _v }),
      includeSchema,
      readonly,
      triggerRender: () => /* @__PURE__ */ import_react11.default.createElement(import_semi_ui6.IconButton, { disabled: readonly, size: "small", icon: /* @__PURE__ */ import_react11.default.createElement(import_semi_icons5.IconSetting, { size: "small" }) })
    }
  );
  return /* @__PURE__ */ import_react11.default.createElement(UIContainer2, { style }, /* @__PURE__ */ import_react11.default.createElement(UIMain, null, renderMain()), /* @__PURE__ */ import_react11.default.createElement(UITrigger, null, renderTrigger()));
}

// src/components/condition-row/index.tsx
var import_react14 = __toESM(require("react"), 1);
var import_semi_ui8 = require("@douyinfe/semi-ui");

// src/components/condition-row/styles.tsx
var import_styled_components4 = __toESM(require("styled-components"), 1);
var UIContainer3 = import_styled_components4.default.div`
  display: flex;
  align-items: center;
  gap: 4px;
`;
var UIOperator = import_styled_components4.default.div``;
var UILeft = import_styled_components4.default.div`
  width: 100%;
`;
var UIRight = import_styled_components4.default.div`
  width: 100%;
`;
var UIValues = import_styled_components4.default.div`
  flex-grow: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 4px;
`;

// src/components/condition-row/hooks/useRule.ts
var import_react12 = require("react");
var import_editor4 = require("@flowgram.ai/editor");

// src/components/condition-row/constants.ts
var rules = {
  string: {
    ["eq" /* EQ */]: "string",
    ["neq" /* NEQ */]: "string",
    ["contains" /* CONTAINS */]: "string",
    ["not_contains" /* NOT_CONTAINS */]: "string",
    ["in" /* IN */]: "array",
    ["nin" /* NIN */]: "array",
    ["is_empty" /* IS_EMPTY */]: "string",
    ["is_not_empty" /* IS_NOT_EMPTY */]: "string"
  },
  number: {
    ["eq" /* EQ */]: "number",
    ["neq" /* NEQ */]: "number",
    ["gt" /* GT */]: "number",
    ["gte" /* GTE */]: "number",
    ["lt" /* LT */]: "number",
    ["lte" /* LTE */]: "number",
    ["in" /* IN */]: "array",
    ["nin" /* NIN */]: "array",
    ["is_empty" /* IS_EMPTY */]: null,
    ["is_not_empty" /* IS_NOT_EMPTY */]: null
  },
  integer: {
    ["eq" /* EQ */]: "number",
    ["neq" /* NEQ */]: "number",
    ["gt" /* GT */]: "number",
    ["gte" /* GTE */]: "number",
    ["lt" /* LT */]: "number",
    ["lte" /* LTE */]: "number",
    ["in" /* IN */]: "array",
    ["nin" /* NIN */]: "array",
    ["is_empty" /* IS_EMPTY */]: null,
    ["is_not_empty" /* IS_NOT_EMPTY */]: null
  },
  boolean: {
    ["eq" /* EQ */]: "boolean",
    ["neq" /* NEQ */]: "boolean",
    ["is_true" /* IS_TRUE */]: null,
    ["is_false" /* IS_FALSE */]: null,
    ["in" /* IN */]: "array",
    ["nin" /* NIN */]: "array",
    ["is_empty" /* IS_EMPTY */]: null,
    ["is_not_empty" /* IS_NOT_EMPTY */]: null
  },
  object: {
    ["is_empty" /* IS_EMPTY */]: null,
    ["is_not_empty" /* IS_NOT_EMPTY */]: null
  },
  array: {
    ["is_empty" /* IS_EMPTY */]: null,
    ["is_not_empty" /* IS_NOT_EMPTY */]: null
  },
  map: {
    ["is_empty" /* IS_EMPTY */]: null,
    ["is_not_empty" /* IS_NOT_EMPTY */]: null
  }
};
var opConfigs = {
  ["eq" /* EQ */]: {
    label: "Equal",
    abbreviation: "="
  },
  ["neq" /* NEQ */]: {
    label: "Not Equal",
    abbreviation: "\u2260"
  },
  ["gt" /* GT */]: {
    label: "Greater Than",
    abbreviation: ">"
  },
  ["gte" /* GTE */]: {
    label: "Greater Than or Equal",
    abbreviation: ">="
  },
  ["lt" /* LT */]: {
    label: "Less Than",
    abbreviation: "<"
  },
  ["lte" /* LTE */]: {
    label: "Less Than or Equal",
    abbreviation: "<="
  },
  ["in" /* IN */]: {
    label: "In",
    abbreviation: "\u2208"
  },
  ["nin" /* NIN */]: {
    label: "Not In",
    abbreviation: "\u2209"
  },
  ["contains" /* CONTAINS */]: {
    label: "Contains",
    abbreviation: "\u2287"
  },
  ["not_contains" /* NOT_CONTAINS */]: {
    label: "Not Contains",
    abbreviation: "\u2289"
  },
  ["is_empty" /* IS_EMPTY */]: {
    label: "Is Empty",
    abbreviation: "=",
    rightDisplay: "Empty"
  },
  ["is_not_empty" /* IS_NOT_EMPTY */]: {
    label: "Is Not Empty",
    abbreviation: "\u2260",
    rightDisplay: "Empty"
  },
  ["is_true" /* IS_TRUE */]: {
    label: "Is True",
    abbreviation: "=",
    rightDisplay: "True"
  },
  ["is_false" /* IS_FALSE */]: {
    label: "Is False",
    abbreviation: "=",
    rightDisplay: "False"
  }
};

// src/utils/format-legacy-refs/index.ts
var import_lodash2 = require("lodash");
function formatLegacyRefOnSubmit(value) {
  if ((0, import_lodash2.isObject)(value)) {
    if (isLegacyFlowRefValueSchema(value)) {
      return formatLegacyRefToNewRef(value);
    }
    return Object.fromEntries(
      Object.entries(value).map(([key, value2]) => [
        key,
        formatLegacyRefOnSubmit(value2)
      ])
    );
  }
  if (Array.isArray(value)) {
    return value.map(formatLegacyRefOnSubmit);
  }
  return value;
}
function formatLegacyRefOnInit(value) {
  if ((0, import_lodash2.isObject)(value)) {
    if (isNewFlowRefValueSchema(value)) {
      return formatNewRefToLegacyRef(value);
    }
    return Object.fromEntries(
      Object.entries(value).map(([key, value2]) => [
        key,
        formatLegacyRefOnInit(value2)
      ])
    );
  }
  if (Array.isArray(value)) {
    return value.map(formatLegacyRefOnInit);
  }
  return value;
}
function isLegacyFlowRefValueSchema(value) {
  return (0, import_lodash2.isObject)(value) && Object.keys(value).length === 2 && value.type === "ref" && typeof value.content === "string";
}
function isNewFlowRefValueSchema(value) {
  return (0, import_lodash2.isObject)(value) && Object.keys(value).length === 2 && value.type === "ref" && Array.isArray(value.content);
}
function formatLegacyRefToNewRef(value) {
  const keyPath = value.content.split(".");
  if (keyPath[1] === "outputs") {
    return {
      type: "ref",
      content: [`${keyPath[0]}.${keyPath[1]}`, ...keyPath.length > 2 ? keyPath.slice(2) : []]
    };
  }
  return {
    type: "ref",
    content: keyPath
  };
}
function formatNewRefToLegacyRef(value) {
  return {
    type: "ref",
    content: value.content.join(".")
  };
}

// src/components/condition-row/hooks/useRule.ts
function useRule(left) {
  const available = (0, import_editor4.useScopeAvailable)();
  const variable = (0, import_react12.useMemo)(() => {
    if (!left) return void 0;
    return available.getByKeyPath(left.content);
  }, [available, left]);
  const rule = (0, import_react12.useMemo)(() => {
    if (!variable) return void 0;
    const schema = JsonSchemaUtils.astToSchema(variable.type, { drilldown: false });
    return rules[schema?.type];
  }, [variable?.type]);
  return { rule };
}

// src/components/condition-row/hooks/useOp.tsx
var import_react13 = __toESM(require("react"), 1);
var import_semi_ui7 = require("@douyinfe/semi-ui");
var import_semi_icons6 = require("@douyinfe/semi-icons");
function useOp({ rule, op, onChange }) {
  const options2 = (0, import_react13.useMemo)(
    () => Object.keys(rule || {}).map((_op) => ({
      ...opConfigs[_op] || {},
      value: _op
    })),
    [rule]
  );
  const opConfig = (0, import_react13.useMemo)(() => opConfigs[op], [op]);
  const renderOpSelect = () => /* @__PURE__ */ import_react13.default.createElement(
    import_semi_ui7.Select,
    {
      style: { height: 22 },
      size: "small",
      value: op,
      optionList: options2,
      onChange: (v) => {
        onChange(v);
      },
      triggerRender: ({ value }) => /* @__PURE__ */ import_react13.default.createElement(import_semi_ui7.Button, { size: "small", disabled: !rule }, opConfig?.abbreviation || /* @__PURE__ */ import_react13.default.createElement(import_semi_icons6.IconChevronDownStroked, { size: "small" }))
    }
  );
  return { renderOpSelect, opConfig };
}

// src/components/condition-row/index.tsx
function ConditionRow({ style, value, onChange, readonly }) {
  const { left, operator, right } = value || {};
  const { rule } = useRule(left);
  const { renderOpSelect, opConfig } = useOp({
    rule,
    op: operator,
    onChange: (v) => onChange({ ...value, operator: v })
  });
  const targetSchema = (0, import_react14.useMemo)(() => {
    const targetType = rule?.[operator] || null;
    return targetType ? { type: targetType, extra: { weak: true } } : null;
  }, [rule, opConfig]);
  return /* @__PURE__ */ import_react14.default.createElement(UIContainer3, { style }, /* @__PURE__ */ import_react14.default.createElement(UIOperator, null, renderOpSelect()), /* @__PURE__ */ import_react14.default.createElement(UIValues, null, /* @__PURE__ */ import_react14.default.createElement(UILeft, null, /* @__PURE__ */ import_react14.default.createElement(
    VariableSelector,
    {
      readonly,
      style: { width: "100%" },
      value: left?.content,
      onChange: (v) => onChange({
        ...value,
        left: {
          type: "ref",
          content: v
        }
      })
    }
  )), /* @__PURE__ */ import_react14.default.createElement(UIRight, null, targetSchema ? /* @__PURE__ */ import_react14.default.createElement(
    DynamicValueInput,
    {
      readonly: readonly || !rule,
      value: right,
      schema: targetSchema,
      onChange: (v) => onChange({ ...value, right: v })
    }
  ) : /* @__PURE__ */ import_react14.default.createElement(import_semi_ui8.Input, { size: "small", disabled: true, value: opConfig?.rightDisplay || "Empty" }))));
}

// src/effects/provide-batch-input/index.ts
var import_editor5 = require("@flowgram.ai/editor");
var provideBatchInputEffect = (0, import_editor5.createEffectFromVariableProvider)({
  private: true,
  parse: (value, ctx) => [
    import_editor5.ASTFactory.createVariableDeclaration({
      key: `${ctx.node.id}_locals`,
      meta: {
        title: (0, import_editor5.getNodeForm)(ctx.node)?.getValueIn("title"),
        icon: ctx.node.getNodeRegistry().info?.icon
      },
      type: import_editor5.ASTFactory.createObject({
        properties: [
          import_editor5.ASTFactory.createProperty({
            key: "item",
            initializer: import_editor5.ASTFactory.createEnumerateExpression({
              enumerateFor: import_editor5.ASTFactory.createKeyPathExpression({
                keyPath: value.content || []
              })
            })
          }),
          import_editor5.ASTFactory.createProperty({
            key: "index",
            type: import_editor5.ASTFactory.createNumber()
          })
        ]
      })
    })
  ]
});

// src/effects/provide-batch-outputs/index.ts
var import_editor6 = require("@flowgram.ai/editor");
var provideBatchOutputsEffect = (0, import_editor6.createEffectFromVariableProvider)({
  private: true,
  parse: (value, ctx) => [
    import_editor6.ASTFactory.createVariableDeclaration({
      key: `${ctx.node.id}`,
      meta: {
        title: (0, import_editor6.getNodeForm)(ctx.node)?.getValueIn("title"),
        icon: ctx.node.getNodeRegistry().info?.icon
      },
      type: import_editor6.ASTFactory.createObject({
        properties: Object.entries(value).map(
          ([_key, value2]) => import_editor6.ASTFactory.createProperty({
            key: _key,
            initializer: import_editor6.ASTFactory.createWrapArrayExpression({
              wrapFor: import_editor6.ASTFactory.createKeyPathExpression({
                keyPath: value2.content || []
              })
            })
          })
        )
      })
    })
  ]
});

// src/effects/auto-rename-ref/index.ts
var import_lodash3 = require("lodash");
var import_editor7 = require("@flowgram.ai/editor");
var autoRenameRefEffect = [
  {
    event: import_editor7.DataEvent.onValueInit,
    effect: (params) => {
      const { context, form, name } = params;
      const renameService = context.node.getService(import_editor7.VariableFieldKeyRenameService);
      const disposable = renameService.onRename(({ before, after }) => {
        const beforeKeyPath = [
          ...before.parentFields.map((_field) => _field.key).reverse(),
          before.key
        ];
        const afterKeyPath = [
          ...after.parentFields.map((_field) => _field.key).reverse(),
          after.key
        ];
        traverseRef(name, form.getValueIn(name), (_drilldownName, _v) => {
          if (isRefMatch(_v, beforeKeyPath)) {
            _v.content = [...afterKeyPath, ...(_v.content || [])?.slice(beforeKeyPath.length)];
            form.setValueIn(_drilldownName, _v);
          }
        });
      });
      return () => {
        disposable.dispose();
      };
    }
  }
];
function isRefMatch(value, targetKeyPath) {
  return targetKeyPath.every((_key, index) => _key === value.content?.[index]);
}
function isRef(value) {
  return value?.type === "ref" && Array.isArray(value?.content) && typeof value?.content[0] === "string";
}
function traverseRef(name, value, cb) {
  if ((0, import_lodash3.isObject)(value)) {
    if (isRef(value)) {
      cb(name, value);
      return;
    }
    Object.entries(value).forEach(([_key, _value]) => {
      traverseRef(`${name}.${_key}`, _value, cb);
    });
    return;
  }
  if ((0, import_lodash3.isArray)(value)) {
    value.forEach((_value, idx) => {
      traverseRef(`${name}[${idx}]`, _value, cb);
    });
    return;
  }
  return;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ArrayIcons,
  BatchVariableSelector,
  ConditionRow,
  ConstantInput,
  DynamicValueInput,
  JsonSchemaEditor,
  JsonSchemaUtils,
  TypeSelector,
  VariableSelector,
  VariableTypeIcons,
  autoRenameRefEffect,
  formatLegacyRefOnInit,
  formatLegacyRefOnSubmit,
  formatLegacyRefToNewRef,
  formatNewRefToLegacyRef,
  getSchemaIcon,
  getTypeSelectValue,
  isLegacyFlowRefValueSchema,
  isNewFlowRefValueSchema,
  parseTypeSelectValue,
  provideBatchInputEffect,
  provideBatchOutputsEffect
});
