// @flowgram.ai/history-node-plugin/dist/esm/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result) __defProp(target, key, result);
  return result;
};

// src/create-history-node-plugin.ts
import { FlowDocument as FlowDocument2 } from "@flowgram.ai/document";
import { bindContributions, definePluginCreator } from "@flowgram.ai/core";
import { HistoryContainerModule, HistoryService, OperationContribution } from "@flowgram.ai/history";

// src/utils/index.ts
import { get } from "lodash";
import { isFormModelV2 } from "@flowgram.ai/node";
import { FlowNodeFormData } from "@flowgram.ai/form-core";
function getFormModelV2(node) {
  if (!node) {
    return void 0;
  }
  const formModel = node?.getData(FlowNodeFormData)?.getFormModel();
  if (!formModel || !isFormModelV2(formModel)) {
    return void 0;
  }
  return formModel;
}
function shouldChangeFormValuesMerge(op, prev, element) {
  if (!prev) {
    return false;
  }
  if (Date.now() - element.getTimestamp() < 500) {
    if (op.type === prev.type && // 相同类型
    op.value?.id === prev.value?.id && // 相同节点
    op.value?.path === prev.value?.path) {
      return {
        type: op.type,
        value: {
          ...op.value,
          value: op.value?.value,
          oldValue: prev.value?.oldValue
        }
      };
    }
    return true;
  }
  return false;
}
function attachFormValuesChange(formModel, node, historyService) {
  formModel.onFormValuesChange((event) => {
    historyService.pushOperation(
      {
        type: "changeFormValues" /* changeFormValues */,
        value: {
          id: node.id,
          path: event.name,
          value: get(event.values, event.name),
          oldValue: get(event.prevValues, event.name)
        }
      },
      { noApply: true }
    );
  });
}

// src/history-node-registers.ts
import { injectable } from "inversify";

// src/operation-metas/change-form-values.ts
import { FlowDocument } from "@flowgram.ai/document";
var changeFormValueOperationMeta = {
  type: "changeFormValues" /* changeFormValues */,
  inverse: (op) => ({
    ...op,
    value: {
      ...op.value,
      value: op.value.oldValue,
      oldValue: op.value.value
    }
  }),
  apply: ({ value: { value, path, id } }, ctx) => {
    const document = ctx.get(FlowDocument);
    const formModel = getFormModelV2(document.getNode(id));
    if (!formModel) {
      return;
    }
    formModel.setValueIn(path, value);
  },
  shouldMerge: shouldChangeFormValuesMerge
};

// src/operation-metas/index.ts
var operationMetas = [changeFormValueOperationMeta];

// src/history-node-registers.ts
var HistoryNodeRegisters = class {
  registerOperationMeta(operationRegistry) {
    operationMetas.forEach((operationMeta) => {
      operationRegistry.registerOperationMeta(operationMeta);
    });
  }
};
HistoryNodeRegisters = __decorateClass([
  injectable()
], HistoryNodeRegisters);

// src/create-history-node-plugin.ts
var createHistoryNodePlugin = definePluginCreator({
  onBind: ({ bind }) => {
    bindContributions(bind, HistoryNodeRegisters, [OperationContribution]);
  },
  onInit: (ctx, _opts) => {
    const document = ctx.get(FlowDocument2);
    const historyService = ctx.get(HistoryService);
    document.onNodeCreate(({ node }) => {
      const formModel = getFormModelV2(node);
      if (!formModel) {
        return;
      }
      attachFormValuesChange(formModel, node, historyService);
    });
  },
  containerModules: [HistoryContainerModule]
});
export {
  createHistoryNodePlugin
};
