// @flowgram.ai/eslint-config/dist/defineConfig.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defineConfig = void 0;
const path_1 = __importDefault(require("path"));
/**
 * Define an ESLint config.
 *
 * @param config ESLint config.
 * @returns ESLint config.
 */
const defineConfig = (config) => {
    const { packageRoot, preset, settings, extends: _extends = [], rules = {}, ...userConfig } = config;
    const defaultRules = {};
    return {
        extends: [path_1.default.resolve(__dirname, `../.eslintrc.${preset}.js`), ..._extends],
        settings: {
            ...settings,
            'import/resolver': {
                typescript: {
                    project: packageRoot,
                },
            },
        },
        rules: {
            ...defaultRules,
            ...rules,
        },
        ...userConfig,
    };
};
exports.defineConfig = defineConfig;
