// @flowgram.ai/free-node-panel-plugin/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result) __defProp(target, key, result);
  return result;
};

// src/index.ts
var src_exports = {};
__export(src_exports, {
  WorkflowNodePanelService: () => WorkflowNodePanelService,
  WorkflowNodePanelUtils: () => WorkflowNodePanelUtils,
  createFreeNodePanelPlugin: () => createFreeNodePanelPlugin
});
module.exports = __toCommonJS(src_exports);

// src/create-plugin.ts
var import_core4 = require("@flowgram.ai/core");

// src/service.ts
var import_inversify = require("inversify");
var import_utils4 = require("@flowgram.ai/utils");
var import_free_layout_core3 = require("@flowgram.ai/free-layout-core");
var import_free_layout_core4 = require("@flowgram.ai/free-layout-core");
var import_free_history_plugin2 = require("@flowgram.ai/free-history-plugin");
var import_core2 = require("@flowgram.ai/core");

// src/utils/wait-node-render.ts
var import_free_layout_core = require("@flowgram.ai/free-layout-core");
var waitNodeRender = async () => {
  await (0, import_free_layout_core.delay)(20);
};

// src/utils/update-sub-nodes-position.ts
var import_free_history_plugin = require("@flowgram.ai/free-history-plugin");
var import_core = require("@flowgram.ai/core");

// src/utils/get-port-box.ts
var import_utils = require("@flowgram.ai/utils");
var import_document = require("@flowgram.ai/document");

// src/utils/is-container.ts
var isContainer = (node) => node?.getNodeMeta().isContainer ?? false;

// src/utils/get-port-box.ts
var getPortBox = (port, offset = { x: 0, y: 0 }) => {
  const node = port.node;
  if (isContainer(node)) {
    const { point } = port;
    if (port.portType === "input") {
      return new import_utils.Rectangle(point.x + offset.x, point.y - 50 + offset.y, 300, 100);
    }
    return new import_utils.Rectangle(point.x - 300, point.y - 50, 300, 100);
  }
  const box = node.getData(import_document.FlowNodeTransformData).bounds;
  return box;
};

// src/utils/update-sub-nodes-position.ts
var updateSubSequentNodesPosition = (params) => {
  const {
    node,
    subsequentNodes,
    fromPort,
    toPort,
    containerNode,
    offset,
    historyService,
    dragService
  } = params;
  if (!offset || !toPort) {
    return;
  }
  const subsequentNodesPositions = subsequentNodes.map((node2) => {
    const nodeTrans2 = node2.getData(import_core.TransformData);
    return {
      x: nodeTrans2.position.x,
      y: nodeTrans2.position.y
    };
  });
  historyService.pushOperation({
    type: import_free_history_plugin.FreeOperationType.dragNodes,
    value: {
      ids: subsequentNodes.map((node2) => node2.id),
      value: subsequentNodesPositions.map((position) => ({
        x: position.x + offset.x,
        y: position.y + offset.y
      })),
      oldValue: subsequentNodesPositions
    }
  });
  const fromBox = getPortBox(fromPort);
  const toBox = getPortBox(toPort, offset);
  const nodeTrans = node.getData(import_core.TransformData);
  let nodePos = {
    x: (fromBox.center.x + toBox.center.x) / 2,
    y: (fromBox.y + toBox.y) / 2
  };
  if (containerNode) {
    nodePos = dragService.adjustSubNodePosition(
      node.flowNodeType,
      containerNode,
      nodePos
    );
  }
  historyService.pushOperation({
    type: import_free_history_plugin.FreeOperationType.dragNodes,
    value: {
      ids: [node.id],
      value: [nodePos],
      oldValue: [
        {
          x: nodeTrans.position.x,
          y: nodeTrans.position.y
        }
      ]
    }
  });
};

// src/utils/sub-position-offset.ts
var import_utils3 = require("@flowgram.ai/utils");
var import_document2 = require("@flowgram.ai/document");

// src/utils/rect-distance.ts
var import_utils2 = require("@flowgram.ai/utils");
var rectDistance = (rectA, rectB) => {
  const distanceX = Math.abs(Math.min(rectA.right, rectB.right) - Math.max(rectA.left, rectB.left));
  const distanceY = Math.abs(Math.min(rectA.bottom, rectB.bottom) - Math.max(rectA.top, rectB.top));
  if (import_utils2.Rectangle.intersects(rectA, rectB)) {
    return {
      x: -distanceX,
      y: -distanceY
    };
  }
  return {
    x: distanceX,
    y: distanceY
  };
};

// src/utils/greater-or-less.ts
var isGreaterThan = (a, b) => {
  if (a === void 0 || b === void 0) {
    return false;
  }
  const EPSILON = 1e-5;
  return a - b > EPSILON;
};
var isLessThan = (a, b) => {
  if (a === void 0 || b === void 0) {
    return false;
  }
  const EPSILON = 1e-5;
  return b - a > EPSILON;
};

// src/utils/sub-position-offset.ts
var subPositionOffset = (params) => {
  const { node, fromPort, toPort, padding } = params;
  const fromBox = getPortBox(fromPort);
  const toBox = getPortBox(toPort);
  const nodeTrans = node.getData(import_document2.FlowNodeTransformData);
  const nodeSize = node.getNodeMeta()?.size ?? {
    width: nodeTrans.bounds.width,
    height: nodeTrans.bounds.height
  };
  const minDistance = {
    x: nodeSize.width + padding.x,
    y: nodeSize.height + padding.y
  };
  const boxDistance = rectDistance(fromBox, toBox);
  const neededOffset = {
    x: isGreaterThan(boxDistance.x, minDistance.x) ? 0 : minDistance.x - boxDistance.x,
    y: isGreaterThan(boxDistance.y, minDistance.y) ? 0 : minDistance.y - boxDistance.y
  };
  if (neededOffset.x === 0 || neededOffset.y === 0) {
    return;
  }
  const intersection = {
    // 这里没有写反，Rectangle内置的算法是反的
    vertical: import_utils3.Rectangle.intersects(fromBox, toBox, "horizontal"),
    horizontal: import_utils3.Rectangle.intersects(fromBox, toBox, "vertical")
  };
  let offsetX = 0;
  let offsetY = 0;
  if (!intersection.horizontal) {
    if (isGreaterThan(toBox.center.y, fromBox.center.y)) {
      offsetY = neededOffset.y;
    } else if (isLessThan(toBox.center.y, fromBox.center.y)) {
      offsetY = -neededOffset.y;
    }
  }
  if (!intersection.vertical) {
    if (isGreaterThan(toBox.center.x, fromBox.center.x)) {
      offsetX = neededOffset.x;
    } else if (isLessThan(toBox.center.x, fromBox.center.x)) {
      offsetX = -neededOffset.x;
    }
  }
  return {
    x: offsetX,
    y: offsetY
  };
};

// src/utils/get-sub-nodes.ts
var getSubsequentNodes = (params) => {
  const { node, linesManager } = params;
  if (isContainer(node)) {
    return [];
  }
  const brothers = node.parent?.blocks ?? [];
  const linkedBrothers = /* @__PURE__ */ new Set();
  const linesMap = /* @__PURE__ */ new Map();
  linesManager.getAllLines().forEach((line) => {
    if (!linesMap.has(line.from.id)) {
      linesMap.set(line.from.id, []);
    }
    if (!line.to?.id || isContainer(line.to)) {
      return;
    }
    linesMap.get(line.from.id)?.push(line.to.id);
  });
  const bfs = (nodeId) => {
    if (linkedBrothers.has(nodeId)) {
      return;
    }
    linkedBrothers.add(nodeId);
    const nextNodes = linesMap.get(nodeId) ?? [];
    nextNodes.forEach(bfs);
  };
  bfs(node.id);
  const subsequentNodes = brothers.filter((node2) => linkedBrothers.has(node2.id));
  return subsequentNodes;
};

// src/utils/sub-nodes-auto-offset.ts
var subNodesAutoOffset = (params) => {
  const {
    node,
    fromPort,
    toPort,
    linesManager,
    historyService,
    dragService,
    containerNode,
    padding = {
      x: 100,
      y: 100
    }
  } = params;
  const subOffset = subPositionOffset({
    node,
    fromPort,
    toPort,
    padding
  });
  const subsequentNodes = getSubsequentNodes({
    node: toPort.node,
    linesManager
  });
  updateSubSequentNodesPosition({
    node,
    subsequentNodes,
    fromPort,
    toPort,
    containerNode,
    offset: subOffset,
    historyService,
    dragService
  });
};

// src/utils/get-container-node.ts
var getContainerNode = (params) => {
  const { fromPort, containerNode } = params;
  if (containerNode) {
    return containerNode;
  }
  const fromNode = fromPort?.node;
  const fromContainer = fromNode?.parent;
  if (isContainer(fromNode)) {
    return fromNode;
  }
  return fromContainer;
};

// src/utils/build-line.ts
var import_free_layout_core2 = require("@flowgram.ai/free-layout-core");
var buildLine = (params) => {
  const { fromPort, node, toPort, linesManager } = params;
  const portsData = node.getData(import_free_layout_core2.WorkflowNodePortsData);
  if (!portsData) {
    return;
  }
  const shouldBuildFromLine = portsData.inputPorts?.length > 0;
  if (fromPort && shouldBuildFromLine) {
    const toTargetPort = portsData.inputPorts[0];
    const isSingleInput = portsData.inputPorts.length === 1;
    linesManager.createLine({
      from: fromPort.node.id,
      fromPort: fromPort.portID,
      to: node.id,
      toPort: isSingleInput ? void 0 : toTargetPort.id
    });
  }
  const shouldBuildToLine = portsData.outputPorts?.length > 0;
  if (toPort && shouldBuildToLine) {
    const fromTargetPort = portsData.outputPorts[0];
    linesManager.createLine({
      from: node.id,
      fromPort: fromTargetPort.portID,
      to: toPort.node.id,
      toPort: toPort.portID
    });
  }
};

// src/utils/adjust-node-position.ts
var adjustNodePosition = (params) => {
  const { nodeType, position, fromPort, toPort, containerNode, document, dragService } = params;
  const register = document.getNodeRegistry(nodeType);
  const size = register?.meta?.size;
  let adjustedPosition = position;
  if (!size) {
    adjustedPosition = position;
  } else if (fromPort && toPort) {
    adjustedPosition = {
      x: position.x,
      y: position.y - size.height / 2
    };
  } else if (fromPort && !toPort) {
    adjustedPosition = {
      x: position.x + size.width / 2,
      y: position.y - size.height / 2
    };
  } else if (!fromPort && toPort) {
    adjustedPosition = {
      x: position.x - size.width / 2,
      y: position.y - size.height / 2
    };
  } else {
    adjustedPosition = position;
  }
  return dragService.adjustSubNodePosition(nodeType, containerNode, adjustedPosition);
};

// src/utils/index.ts
var WorkflowNodePanelUtils = {
  adjustNodePosition,
  buildLine,
  getPortBox,
  getSubsequentNodes,
  getContainerNode,
  rectDistance,
  subNodesAutoOffset,
  subPositionOffset,
  updateSubSequentNodesPosition,
  waitNodeRender
};

// src/service.ts
var WorkflowNodePanelService = class {
  constructor() {
    this.toDispose = new import_utils4.DisposableCollection();
    this.callNodePanel = async () => void 0;
  }
  /** 销毁 */
  dispose() {
    this.toDispose.dispose();
  }
  setCallNodePanel(callNodePanel) {
    this.callNodePanel = callNodePanel;
  }
  /** 唤起节点面板 */
  async call(callParams) {
    const {
      panelPosition,
      fromPort,
      enableMultiAdd = false,
      panelProps = {},
      containerNode,
      afterAddNode
    } = callParams;
    if (!panelPosition || this.playgroundConfig.readonly) {
      return;
    }
    const nodes = [];
    return new Promise((resolve) => {
      this.callNodePanel({
        position: panelPosition,
        enableMultiAdd,
        panelProps,
        containerNode: WorkflowNodePanelUtils.getContainerNode({
          fromPort,
          containerNode
        }),
        onSelect: async (panelParams) => {
          const node = await this.addNode(callParams, panelParams);
          afterAddNode?.(node);
          if (!enableMultiAdd) {
            resolve(node);
          } else if (node) {
            nodes.push(node);
          }
        },
        onClose: () => {
          resolve(enableMultiAdd ? nodes : void 0);
        }
      });
    });
  }
  /**
   * 唤起单选面板
   */
  async singleSelectNodePanel(params) {
    return new Promise((resolve) => {
      this.callNodePanel({
        ...params,
        enableMultiAdd: false,
        onSelect: async (panelParams) => {
          resolve(panelParams);
        },
        onClose: () => {
          resolve(void 0);
        }
      });
    });
  }
  /** 添加节点 */
  async addNode(callParams, panelParams) {
    const {
      panelPosition,
      fromPort,
      toPort,
      canAddNode,
      autoOffsetPadding = {
        x: 100,
        y: 100
      },
      enableBuildLine = false,
      enableSelectPosition = false,
      enableAutoOffset = false,
      enableDragNode = false
    } = callParams;
    if (!panelPosition || !panelParams) {
      return;
    }
    const { nodeType, selectEvent, nodeJSON } = panelParams;
    const containerNode = WorkflowNodePanelUtils.getContainerNode({
      fromPort,
      containerNode: callParams.containerNode
    });
    if (canAddNode) {
      const canAdd = canAddNode({ nodeType, containerNode });
      if (!canAdd) {
        return;
      }
    }
    const selectPosition = this.playgroundConfig.getPosFromMouseEvent(selectEvent);
    const nodePosition = callParams.customPosition ? callParams.customPosition({ nodeType, selectPosition }) : WorkflowNodePanelUtils.adjustNodePosition({
      nodeType,
      position: enableSelectPosition ? selectPosition : panelPosition,
      fromPort,
      toPort,
      containerNode,
      document: this.document,
      dragService: this.dragService
    });
    const node = this.document.createWorkflowNodeByType(
      nodeType,
      nodePosition,
      nodeJSON ?? {},
      containerNode?.id
    );
    if (!node) {
      return;
    }
    if (enableAutoOffset && fromPort && toPort) {
      WorkflowNodePanelUtils.subNodesAutoOffset({
        node,
        fromPort,
        toPort,
        padding: autoOffsetPadding,
        containerNode,
        historyService: this.historyService,
        dragService: this.dragService,
        linesManager: this.linesManager
      });
    }
    if (!enableBuildLine && !enableDragNode) {
      return node;
    }
    await WorkflowNodePanelUtils.waitNodeRender();
    if (enableBuildLine) {
      WorkflowNodePanelUtils.buildLine({
        fromPort,
        node,
        toPort,
        linesManager: this.linesManager
      });
    }
    if (enableDragNode) {
      this.selectService.selectNode(node);
      this.dragService.startDragSelectedNodes(selectEvent);
    }
    return node;
  }
};
__decorateClass([
  (0, import_inversify.inject)(import_free_layout_core3.WorkflowDocument)
], WorkflowNodePanelService.prototype, "document", 2);
__decorateClass([
  (0, import_inversify.inject)(import_free_layout_core3.WorkflowDragService)
], WorkflowNodePanelService.prototype, "dragService", 2);
__decorateClass([
  (0, import_inversify.inject)(import_free_layout_core4.WorkflowSelectService)
], WorkflowNodePanelService.prototype, "selectService", 2);
__decorateClass([
  (0, import_inversify.inject)(import_free_layout_core3.WorkflowLinesManager)
], WorkflowNodePanelService.prototype, "linesManager", 2);
__decorateClass([
  (0, import_inversify.inject)(import_core2.PlaygroundConfigEntity)
], WorkflowNodePanelService.prototype, "playgroundConfig", 2);
__decorateClass([
  (0, import_inversify.inject)(import_free_history_plugin2.HistoryService)
], WorkflowNodePanelService.prototype, "historyService", 2);
WorkflowNodePanelService = __decorateClass([
  (0, import_inversify.injectable)()
], WorkflowNodePanelService);

// src/layer.tsx
var import_react = __toESM(require("react"));
var import_inversify2 = require("inversify");
var import_utils6 = require("@flowgram.ai/utils");
var import_free_layout_core5 = require("@flowgram.ai/free-layout-core");
var import_core3 = require("@flowgram.ai/core");
var WorkflowNodePanelLayer = class extends import_core3.Layer {
  constructor() {
    super();
    this.node = import_utils6.domUtils.createDivWithClass("gedit-playground-layer gedit-node-panel-layer");
    this.node.style.zIndex = "9999";
    this.renderList = /* @__PURE__ */ new Map();
  }
  onReady() {
    this.service.setCallNodePanel(this.call.bind(this));
  }
  onZoom(zoom) {
    this.node.style.transform = `scale(${zoom})`;
  }
  render() {
    const NodePanelRender = this.options.renderer;
    return /* @__PURE__ */ import_react.default.createElement(import_react.default.Fragment, null, Array.from(this.renderList.keys()).map((taskId) => {
      const renderProps = this.renderList.get(taskId);
      return /* @__PURE__ */ import_react.default.createElement(NodePanelRender, { key: taskId, ...renderProps });
    }));
  }
  async call(params) {
    const taskId = (0, import_free_layout_core5.nanoid)();
    const { onSelect, onClose, enableMultiAdd = false, panelProps = {} } = params;
    return new Promise((resolve) => {
      const unmount = () => {
        this.renderList.delete(taskId);
        this.render();
        resolve();
      };
      const handleClose = () => {
        unmount();
        onClose();
      };
      const handleSelect = (params2) => {
        onSelect(params2);
        if (!enableMultiAdd) {
          unmount();
        }
      };
      const renderProps = {
        ...params,
        panelProps,
        onSelect: handleSelect,
        onClose: handleClose
      };
      this.renderList.set(taskId, renderProps);
      this.render();
    });
  }
};
WorkflowNodePanelLayer.type = "WorkflowNodePanelLayer";
__decorateClass([
  (0, import_inversify2.inject)(WorkflowNodePanelService)
], WorkflowNodePanelLayer.prototype, "service", 2);

// src/create-plugin.ts
var createFreeNodePanelPlugin = (0, import_core4.definePluginCreator)({
  onBind({ bind }) {
    bind(WorkflowNodePanelService).toSelf().inSingletonScope();
  },
  onInit: (ctx, opts) => {
    ctx.playground.registerLayer(WorkflowNodePanelLayer, {
      renderer: opts.renderer
    });
  },
  onDispose: (ctx) => {
    const nodePanelService = ctx.get(WorkflowNodePanelService);
    nodePanelService.dispose();
  }
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  WorkflowNodePanelService,
  WorkflowNodePanelUtils,
  createFreeNodePanelPlugin
});
