// @flowgram.ai/redux-devtool-plugin/dist/esm/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result) __defProp(target, key, result);
  return result;
};

// src/create-redux-devtool-plugin.ts
import { definePluginCreator } from "@flowgram.ai/core";

// src/connectors/ecs-connector.ts
import { inject, injectable as injectable2 } from "inversify";
import { EntityManager } from "@flowgram.ai/core";

// src/connectors/base.ts
import { injectable, postConstruct } from "inversify";
var BaseConnector = class {
  constructor() {
    this.devTools = window.__REDUX_DEVTOOLS_EXTENSION__?.connect({
      name: this.getName()
    });
  }
  init() {
    if (this.devTools) {
      this.devTools.init(this.getState());
      this.devTools.subscribe((_msg) => {
        if (_msg?.payload?.type === "COMMIT") {
          this.devTools.init(this.getState());
        }
      });
      this.onInit();
    }
  }
  send(action, state) {
    if (this.devTools) {
      this.devTools.send(action, state || this.getState());
    }
  }
};
__decorateClass([
  postConstruct()
], BaseConnector.prototype, "init", 1);
BaseConnector = __decorateClass([
  injectable()
], BaseConnector);

// src/connectors/ecs-connector.ts
var ECSConnector = class extends BaseConnector {
  getName() {
    return "@flowgram.ai/EntityManager";
  }
  getState() {
    return this.entityManager.storeState({ configOnly: false });
  }
  onInit() {
    this.entityManager.onEntityLifeCycleChange((action) => {
      this.send(`${action.type}/${action.entity.type}/${action.entity.id}`);
    });
  }
};
__decorateClass([
  inject(EntityManager)
], ECSConnector.prototype, "entityManager", 2);
ECSConnector = __decorateClass([
  injectable2()
], ECSConnector);

// src/connectors/variable-connector.ts
import { inject as inject2, injectable as injectable3 } from "inversify";
import { VariableEngine } from "@flowgram.ai/variable-core";
var VariableConnector = class extends BaseConnector {
  constructor() {
    super(...arguments);
    /**
     * 缓存变量状态
     */
    this.scopes = {};
  }
  getName() {
    return "@flowgram.ai/VariableEngine";
  }
  getState() {
    return { scopes: this.scopes, variables: this.variableEngine.globalVariableTable.variables };
  }
  getScopeState(scope) {
    return {
      ast: scope?.ast.toJSON(),
      output: scope.output.variables,
      available: scope.available.variables
    };
  }
  onInit() {
    this.variableEngine.onScopeChange((action) => {
      const { scope, type } = action;
      if (type === "delete") {
        delete this.scopes[String(scope.id)];
      } else {
        this.scopes = {
          ...this.scopes,
          [scope.id]: this.getScopeState(scope)
        };
      }
      this.send(`${type}/${String(scope.id)}`);
    });
  }
};
__decorateClass([
  inject2(VariableEngine)
], VariableConnector.prototype, "variableEngine", 2);
VariableConnector = __decorateClass([
  injectable3()
], VariableConnector);

// src/create-redux-devtool-plugin.ts
var createReduxDevToolPlugin = definePluginCreator({
  onBind({ bind }, opts) {
    const { enable } = opts;
    if (!enable) {
      return;
    }
    bind(ECSConnector).toSelf().inSingletonScope();
    bind(VariableConnector).toSelf().inSingletonScope();
  },
  onInit(ctx, opts) {
    const { enable, ecs = true, variable = false } = opts;
    if (!enable) {
      return;
    }
    if (ecs) {
      ctx.get(ECSConnector);
    }
    if (variable) {
      ctx.get(VariableConnector);
    }
  }
});
export {
  ECSConnector,
  createReduxDevToolPlugin
};
