// @flowgram.ai/free-layout-core/dist/typings/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/typings/index.ts
var typings_exports = {};
__export(typings_exports, {
  LineColors: () => LineColors,
  LineType: () => LineType,
  URLParams: () => URLParams,
  WorkflowContentChangeType: () => WorkflowContentChangeType,
  WorkflowOperationBaseService: () => WorkflowOperationBaseService
});
module.exports = __toCommonJS(typings_exports);

// src/typings/workflow-json.ts
var WorkflowContentChangeType = /* @__PURE__ */ ((WorkflowContentChangeType2) => {
  WorkflowContentChangeType2["ADD_NODE"] = "ADD_NODE";
  WorkflowContentChangeType2["DELETE_NODE"] = "DELETE_NODE";
  WorkflowContentChangeType2["MOVE_NODE"] = "MOVE_NODE";
  WorkflowContentChangeType2["NODE_DATA_CHANGE"] = "NODE_DATA_CHANGE";
  WorkflowContentChangeType2["ADD_LINE"] = "ADD_LINE";
  WorkflowContentChangeType2["DELETE_LINE"] = "DELETE_LINE";
  WorkflowContentChangeType2["META_CHANGE"] = "META_CHANGE";
  return WorkflowContentChangeType2;
})(WorkflowContentChangeType || {});

// src/typings/workflow-line.ts
var LineType = /* @__PURE__ */ ((LineType2) => {
  LineType2[LineType2["BEZIER"] = 0] = "BEZIER";
  LineType2[LineType2["LINE_CHART"] = 1] = "LINE_CHART";
  return LineType2;
})(LineType || {});
var LineColors = /* @__PURE__ */ ((LineColors2) => {
  LineColors2["HIDDEN"] = "transparent";
  LineColors2["DEFUALT"] = "#4d53e8";
  LineColors2["DRAWING"] = "#5DD6E3";
  LineColors2["HOVER"] = "#37d0ff";
  LineColors2["ERROR"] = "red";
  return LineColors2;
})(LineColors || {});

// src/typings/workflow-operation.ts
var WorkflowOperationBaseService = Symbol("WorkflowOperationBaseService");

// src/typings/index.ts
var URLParams = Symbol("");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LineColors,
  LineType,
  URLParams,
  WorkflowContentChangeType,
  WorkflowOperationBaseService
});
