// @flowgram.ai/free-layout-core/dist/typings/workflow-json.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/typings/workflow-json.ts
var workflow_json_exports = {};
__export(workflow_json_exports, {
  WorkflowContentChangeType: () => WorkflowContentChangeType
});
module.exports = __toCommonJS(workflow_json_exports);
var WorkflowContentChangeType = /* @__PURE__ */ ((WorkflowContentChangeType2) => {
  WorkflowContentChangeType2["ADD_NODE"] = "ADD_NODE";
  WorkflowContentChangeType2["DELETE_NODE"] = "DELETE_NODE";
  WorkflowContentChangeType2["MOVE_NODE"] = "MOVE_NODE";
  WorkflowContentChangeType2["NODE_DATA_CHANGE"] = "NODE_DATA_CHANGE";
  WorkflowContentChangeType2["ADD_LINE"] = "ADD_LINE";
  WorkflowContentChangeType2["DELETE_LINE"] = "DELETE_LINE";
  WorkflowContentChangeType2["META_CHANGE"] = "META_CHANGE";
  return WorkflowContentChangeType2;
})(WorkflowContentChangeType || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  WorkflowContentChangeType
});
