// @flowgram.ai/free-layout-core/dist/typings/workflow-line.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/typings/workflow-line.ts
var workflow_line_exports = {};
__export(workflow_line_exports, {
  LineColors: () => LineColors,
  LineType: () => LineType
});
module.exports = __toCommonJS(workflow_line_exports);
var LineType = /* @__PURE__ */ ((LineType2) => {
  LineType2[LineType2["BEZIER"] = 0] = "BEZIER";
  LineType2[LineType2["LINE_CHART"] = 1] = "LINE_CHART";
  return LineType2;
})(LineType || {});
var LineColors = /* @__PURE__ */ ((LineColors2) => {
  LineColors2["HIDDEN"] = "transparent";
  LineColors2["DEFUALT"] = "#4d53e8";
  LineColors2["DRAWING"] = "#5DD6E3";
  LineColors2["HOVER"] = "#37d0ff";
  LineColors2["ERROR"] = "red";
  return LineColors2;
})(LineColors || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LineColors,
  LineType
});
